<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Aberrant splice junction analysis with associated mutations • splice2neo</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Aberrant splice junction analysis with associated mutations">
<meta property="og:description" content="This package provides functions for analyzing alternative splicing and aberrant splice junctions and their creation from or association with somatic mutations. It integrates the output of several tools that predict splicing effects from mutation or RNA-seq data into a unified splice junction format. Splice junctions can be annotated with affected transcript sequences, CDS, and resulting peptide sequences.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">splice2neo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.6.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/splice2neo_workflow.html">Splice2neo example analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/TRON-Bioinformatics/splice2neo/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="splice2neo">splice2neo<a class="anchor" aria-label="anchor" href="#splice2neo"></a>
</h1></div>
<!-- badges: start -->

<p>Documentation: <a href="https://tron-bioinformatics.github.io/splice2neo/" class="uri">https://tron-bioinformatics.github.io/splice2neo/</a></p>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This package provides functions for the analysis of splice junctions and their association with somatic mutations. It integrates the output of several tools that predict splicing effects from mutations or detect expressed splice junctions from RNA-seq data into a standardized splice junction format based on genomic coordinates. Users can filter detected splice junctions against canonical splice junctions and annotate them with affected transcript sequences, CDS, and resulting peptide sequences. Splice2neo supports splice events from alternative 3’/5’ splice sites, exons skipping, intron retentions, exitrons, and mutually exclusive exons. Integrating splice2neo functions and detection rules based on splice effect scores and RNA-seq support facilitates the identification of mutation-associated splice junctions that are tumor-specific and can encode neoantigen candidates.</p>
<p>Currently, splice2neo supports the output data from the following tools:</p>
<ul>
<li>Mutation effect prediction tools:
<ul>
<li><a href="https://github.com/Illumina/SpliceAI" class="external-link">SpliceAI</a></li>
<li><a href="https://github.com/YStrauch/CI-SpliceAI__Annotation" class="external-link">CI-SpliceAI</a></li>
<li><a href="https://github.com/gagneurlab/MMSplice_MTSplice" class="external-link">MMsplice</a></li>
<li><a href="https://github.com/tkzeng/Pangolin" class="external-link">Pangolin</a></li>
</ul>
</li>
<li>RNA-seq-based splice junction detection tools:
<ul>
<li><a href="https://github.com/davidaknowles/leafcutter" class="external-link">LeafCutter</a></li>
<li><a href="https://github.com/griffithlab/regtools" class="external-link">RegTools</a></li>
<li><a href="https://github.com/ratschlab/spladder" class="external-link">SplAdder</a></li>
<li><a href="https://github.com/RitchieLabIGH/IRFinder" class="external-link">IRfinder</a></li>
<li><a href="https://github.com/comprna/SUPPA" class="external-link">SUPPA2</a></li>
<li><a href="https://github.com/alexdobin/STAR" class="external-link">STAR</a></li>
<li><a href="https://github.com/gpertea/stringtie" class="external-link">StringTie</a></li>
</ul>
</li>
</ul>
<p>For a more detailed description and a full example workflow see the <a href="https://tron-bioinformatics.github.io/splice2neo/articles/splice2neo_workflow.html">vignette</a></p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The R package splice2neo is not yet on <a href="https://CRAN.R-project.org" class="external-link">CRAN</a> or <a href="https://www.bioconductor.org/" class="external-link">Bioconductor</a>. Therefore, you need to install splice2neo from <a href="https://github.com/TRON-Bioinformatics/splice2neo" class="external-link">github</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"remotes"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"TRON-Bioinformatics/splice2neo"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="example-usage">Example usage<a class="anchor" aria-label="anchor" href="#example-usage"></a>
</h2>
<p>This is a basic example of how to use some functions.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/TRON-Bioinformatics/splice2neo" class="external-link">splice2neo</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load human genome reference sequence</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">bsg</span> <span class="op">&lt;-</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg19</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html" class="external-link">BSgenome.Hsapiens.UCSC.hg19</a></span></span></code></pre></div>
<div class="section level3">
<h3 id="id_1-example-data">1 Example data<a class="anchor" aria-label="anchor" href="#id_1-example-data"></a>
</h3>
<p>We start with some example splice junctions provided with the package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">junc_df</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  junc_id <span class="op">=</span> <span class="va">toy_junc_id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">6</span>, <span class="fl">10</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">junc_df</span></span>
<span><span class="co">#&gt; # A tibble: 3 × 1</span></span>
<span><span class="co">#&gt;   junc_id                   </span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                     </span></span>
<span><span class="co">#&gt; 1 chr2:152389996-152392205:-</span></span>
<span><span class="co">#&gt; 2 chr2:179415981-179416357:-</span></span>
<span><span class="co">#&gt; 3 chr2:179446225-179446226:-</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="id_2-add-transcripts">2 Add transcripts<a class="anchor" aria-label="anchor" href="#id_2-add-transcripts"></a>
</h3>
<p>Next, we find the transcripts which are in the same genomic region as the splice junction and that may be affected by the junction.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">junc_df</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="reference/add_tx.html">add_tx</a></span><span class="op">(</span><span class="va">toy_transcripts</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 3</span></span>
<span><span class="co">#&gt;   junc_id                    tx_id           tx_lst      </span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                      &lt;chr&gt;           &lt;named list&gt;</span></span>
<span><span class="co">#&gt; 1 chr2:152389996-152392205:- ENST00000409198 &lt;GRanges&gt;   </span></span>
<span><span class="co">#&gt; 2 chr2:152389996-152392205:- ENST00000172853 &lt;GRanges&gt;   </span></span>
<span><span class="co">#&gt; 3 chr2:152389996-152392205:- ENST00000397345 &lt;GRanges&gt;   </span></span>
<span><span class="co">#&gt; 4 chr2:152389996-152392205:- ENST00000427231 &lt;GRanges&gt;   </span></span>
<span><span class="co">#&gt; 5 chr2:152389996-152392205:- ENST00000618972 &lt;GRanges&gt;   </span></span>
<span><span class="co">#&gt; 6 chr2:152389996-152392205:- ENST00000413693 &lt;GRanges&gt;</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="id_3-modify-transcripts-with-junctions">3 Modify transcripts with junctions<a class="anchor" aria-label="anchor" href="#id_3-modify-transcripts-with-junctions"></a>
</h3>
<p>We modify the canonical transcripts by introducing the splice junctions. Then we add the transcript sequence in a fixed-sized window around the junction positions, the context sequence.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">toy_junc_df</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 2</span></span>
<span><span class="co">#&gt;   junc_id                    tx_id          </span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                      &lt;chr&gt;          </span></span>
<span><span class="co">#&gt; 1 chr2:152389996-152392205:- ENST00000409198</span></span>
<span><span class="co">#&gt; 2 chr2:152389996-152390729:- ENST00000409198</span></span>
<span><span class="co">#&gt; 3 chr2:152389955-152389956:- ENST00000397345</span></span>
<span><span class="co">#&gt; 4 chr2:152388410-152392205:- ENST00000409198</span></span>
<span><span class="co">#&gt; 5 chr2:152388410-152390729:- ENST00000409198</span></span>
<span><span class="co">#&gt; 6 chr2:179415981-179416357:- ENST00000342992</span></span>
<span></span>
<span></span>
<span><span class="va">toy_junc_df</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="reference/add_context_seq.html">add_context_seq</a></span><span class="op">(</span>transcripts <span class="op">=</span> <span class="va">toy_transcripts</span>, size <span class="op">=</span> <span class="fl">400</span>, bsg <span class="op">=</span> <span class="va">bsg</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 8</span></span>
<span><span class="co">#&gt;   junc_id       tx_id tx_mod_id junc_pos_tx cts_seq cts_junc_pos cts_size cts_id</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;         &lt;chr&gt; &lt;chr&gt;           &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;           &lt;int&gt; &lt;chr&gt; </span></span>
<span><span class="co">#&gt; 1 chr2:1523899… ENST… ENST0000…       16412 AAGAAG… 200               400 90bfc…</span></span>
<span><span class="co">#&gt; 2 chr2:1523899… ENST… ENST0000…       16517 AAGAAG… 200               400 26f77…</span></span>
<span><span class="co">#&gt; 3 chr2:1523899… ENST… ENST0000…       21620 AAGAAG… 0,200,1745,…     1945 f1f2c…</span></span>
<span><span class="co">#&gt; 4 chr2:1523884… ENST… ENST0000…       16412 AAGAAG… 200               400 d4f9e…</span></span>
<span><span class="co">#&gt; 5 chr2:1523884… ENST… ENST0000…       16517 AAGAAG… 200               400 c715a…</span></span>
<span><span class="co">#&gt; 6 chr2:1794159… ENST… ENST0000…       83789 TGGATT… 200               400 0128d…</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="id_4-annotate-peptide-sequence">4 Annotate peptide sequence<a class="anchor" aria-label="anchor" href="#id_4-annotate-peptide-sequence"></a>
</h3>
<p>Here, we use the splice junctions to modify the coding sequences (CDS) of the reference transcripts. The resulting CDS sequences are translated into protein sequence and further annotated with the peptide around the junction, the relative position of the splice junction in the peptide, and the location of the junction in an open reading frame (ORF).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">toy_junc_df</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  </span>
<span>  <span class="co"># add peptide sequence</span></span>
<span>  <span class="fu"><a href="reference/add_peptide.html">add_peptide</a></span><span class="op">(</span>cds<span class="op">=</span><span class="va">toy_cds</span>, flanking_size <span class="op">=</span> <span class="fl">13</span>, bsg <span class="op">=</span> <span class="va">bsg</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span></span>
<span>  <span class="co"># select subset of columns</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">junc_id</span>, <span class="va">peptide_context</span>, <span class="va">junc_in_orf</span>, <span class="va">cds_description</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 4</span></span>
<span><span class="co">#&gt;   junc_id                    peptide_context         junc_in_orf cds_description</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                      &lt;chr&gt;                   &lt;lgl&gt;       &lt;chr&gt;          </span></span>
<span><span class="co">#&gt; 1 chr2:152389996-152392205:- NRHFKYATQLMNEIC         TRUE        mutated cds    </span></span>
<span><span class="co">#&gt; 2 chr2:152389996-152390729:- HLLAKTAGDQISQIC         TRUE        mutated cds    </span></span>
<span><span class="co">#&gt; 3 chr2:152389955-152389956:- MLTALYNSHMWSQVMSDGM     TRUE        mutated cds    </span></span>
<span><span class="co">#&gt; 4 chr2:152388410-152392205:- NRHFKYATQLMNEIKYRKNYEK… TRUE        mutated cds    </span></span>
<span><span class="co">#&gt; 5 chr2:152388410-152390729:- &lt;NA&gt;                    TRUE        truncated cds  </span></span>
<span><span class="co">#&gt; 6 chr2:179415981-179416357:- SDPSKFTLAVSPVAGTPDYIDV… TRUE        mutated cds</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="issues">Issues<a class="anchor" aria-label="anchor" href="#issues"></a>
</h2>
<p>Please report issues here: <a href="https://github.com/TRON-Bioinformatics/splice2neo/issues" class="external-link uri">https://github.com/TRON-Bioinformatics/splice2neo/issues</a></p>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>Lang, Franziska, Patrick Sorn, Martin Suchan, Alina Henrich, Christian Albrecht, Nina Koehl, Aline Beicht, et al. 2023. “Prediction of Tumor-Specific Splicing from Somatic Mutations as a Source of Neoantigen Candidates.” bioRxiv. <a href="https://doi.org/10.1101/2023.06.27.546494" class="external-link uri">https://doi.org/10.1101/2023.06.27.546494</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/TRON-Bioinformatics/splice2neo/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing splice2neo</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Jonas Ibn-Salem <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-9886-113X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Franziska Lang <br><small class="roles"> Author </small>  </li>
<li>Michelle Cervante <br><small class="roles"> Author </small>  </li>
<li>Johannes Hausmann <br><small class="roles"> Author </small>  </li>
<li>Patrick Sorn <br><small class="roles"> Author </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/TRON-Bioinformatics/splice2neo/actions" class="external-link"><img src="https://github.com/TRON-Bioinformatics/splice2neo/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://codecov.io/gh/TRON-Bioinformatics/splice2neo?branch=master" class="external-link"><img src="https://codecov.io/gh/TRON-Bioinformatics/splice2neo/branch/master/graph/badge.svg" alt="Codecov test coverage"></a></li>
<li><a href="https://github.com/TRON-Bioinformatics/splice2neo" class="external-link"><img src="https://img.shields.io/badge/devel%20version-0.6.6-blue.svg"></a></li>
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="https://img.shields.io/badge/lifecycle-experimental-blue.svg"></a></li>
<li><a href="https://github.com/TRON-Bioinformatics/splice2neo/commits/dev" class="external-link"><img src="https://img.shields.io/github/last-commit/TRON-Bioinformatics/splice2neo.svg"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Jonas Ibn-Salem, Franziska Lang, Michelle Cervante, Johannes Hausmann, Patrick Sorn.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
