<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Aberrant splice junction analysis with associated mutations • splice2neo</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Aberrant splice junction analysis with associated mutations">
<meta property="og:description" content="This package provides functions for the analysis of alternative or 
 aberrant splicing junctions and their creation from or association with 
 somatic mutations. It integrates the output of several tools which predict 
 splicing effects from mutation or RNA-seq data into a common splice junction 
 format based on genomic coordinates. Splice junctions can be annotated with 
 affected transcript sequences, CDS, and resulting peptide sequences.  ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">splice2neo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/TRON-Bioinformatics/splice2neo/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="splice2neo">splice2neo<a class="anchor" aria-label="anchor" href="#splice2neo"></a>
</h1></div>
<!-- badges: start -->

<ol style="list-style-type: decimal">
<li><a href="##1-Overview">Overview</a></li>
<li><a href="##2-Installation">Installation</a></li>
<li><a href="##3-Example">Example</a></li>
<li><a href="##4-Dummy-workflow">Dummy workflow</a></li>
<li><a href="##5-Building-the-transcript%20database">Building the transcript database</a></li>
</ol>
<div class="section level2">
<h2 id="1-overview">1 Overview<a class="anchor" aria-label="anchor" href="#1-overview"></a>
</h2>
<p>This package provides functions for the analysis of alternative splicing junctions and their association with somatic mutations. It integrates the output of several tools which predict splicing effects from mutations or which detect expressed splice junctions from RNA-seq data into a standardized splice junction format based on genomic coordinates. Detected splice junctions can be filtered against canonical ones and annotated with affected transcript sequences, CDS, and resulting peptide sequences. The resulting tumor-specific splice junctions can encode neoantigens.</p>
<p>Splice2neo currently supports events from alternative 3’/5’ splice sites, exons skipping, intron retentions, exitrons and mutually exclusive exons.</p>
<p>In general, a splice junction is defined by the format: <code>chr:start-end:strand</code> .<br>
Intron retentions are a special case. Here, the junctions is defined by the exon-intron or intron-exon boundary, following the format <code>chr:pos-(pos+1):strand</code>.<br>
Two junctions relating to the same intron retention event event will generate the same transcript and peptide context sequence.</p>
<p>Website: <a href="https://tron-bioinformatics.github.io/splice2neo/" class="uri">https://tron-bioinformatics.github.io/splice2neo/</a></p>
<p><img src="reference/figures/README-unnamed-chunk-2-1.png" width="55%"></p>
<p><img src="reference/figures/README-unnamed-chunk-3-1.png" width="55%"></p>
</div>
<div class="section level2">
<h2 id="2-installation">2 Installation<a class="anchor" aria-label="anchor" href="#2-installation"></a>
</h2>
<p>This R package is not yet on <a href="https://CRAN.R-project.org" class="external-link">CRAN</a> or <a href="https://www.bioconductor.org/" class="external-link">Bioconductor</a>. Therefore, you need to install it from this repository.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## install.packages("remotes")</span>
<span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"TRON-Bioinformatics/splice2neo"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="3-example">3 Example<a class="anchor" aria-label="anchor" href="#3-example"></a>
</h2>
<p>This is a basic example of how to use some functions.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/TRON-Bioinformatics/splice2neo" class="external-link">splice2neo</a></span><span class="op">)</span>

<span class="co"># load human genome reference sequence</span>
<span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">bsg</span> <span class="op">&lt;-</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg19</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html" class="external-link">BSgenome.Hsapiens.UCSC.hg19</a></span></code></pre></div>
<div class="section level3">
<h3 id="31-example-data">3.1 Example data<a class="anchor" aria-label="anchor" href="#31-example-data"></a>
</h3>
<p>We start with some example splice junctions provided with the package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">junc_df</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>
  junc_id <span class="op">=</span> <span class="va">toy_junc_id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">6</span>, <span class="fl">10</span><span class="op">)</span><span class="op">]</span>
<span class="op">)</span>

<span class="va">junc_df</span>
<span class="co">#&gt; # A tibble: 3 × 1</span>
<span class="co">#&gt;   junc_id                   </span>
<span class="co">#&gt;   &lt;chr&gt;                     </span>
<span class="co">#&gt; 1 chr2:152389996-152392205:-</span>
<span class="co">#&gt; 2 chr2:179415981-179416357:-</span>
<span class="co">#&gt; 3 chr2:179446225-179446226:-</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="32-add-transcripts">3.2 Add transcripts<a class="anchor" aria-label="anchor" href="#32-add-transcripts"></a>
</h3>
<p>Next, we find the transcripts which are in the same genomic region as the splice junction and that may be affected by the junction.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">junc_df</span> <span class="op">&lt;-</span> <span class="va">junc_df</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> 
  <span class="fu"><a href="reference/add_tx.html">add_tx</a></span><span class="op">(</span><span class="va">toy_transcripts</span><span class="op">)</span>

<span class="va">junc_df</span>
<span class="co">#&gt; # A tibble: 21 × 3</span>
<span class="co">#&gt;    junc_id                    tx_id           tx_lst      </span>
<span class="co">#&gt;    &lt;chr&gt;                      &lt;chr&gt;           &lt;named list&gt;</span>
<span class="co">#&gt;  1 chr2:152389996-152392205:- ENST00000409198 &lt;GRanges&gt;   </span>
<span class="co">#&gt;  2 chr2:152389996-152392205:- ENST00000172853 &lt;GRanges&gt;   </span>
<span class="co">#&gt;  3 chr2:152389996-152392205:- ENST00000397345 &lt;GRanges&gt;   </span>
<span class="co">#&gt;  4 chr2:152389996-152392205:- ENST00000427231 &lt;GRanges&gt;   </span>
<span class="co">#&gt;  5 chr2:152389996-152392205:- ENST00000618972 &lt;GRanges&gt;   </span>
<span class="co">#&gt;  6 chr2:152389996-152392205:- ENST00000413693 &lt;GRanges&gt;   </span>
<span class="co">#&gt;  7 chr2:152389996-152392205:- ENST00000603639 &lt;GRanges&gt;   </span>
<span class="co">#&gt;  8 chr2:152389996-152392205:- ENST00000604864 &lt;GRanges&gt;   </span>
<span class="co">#&gt;  9 chr2:152389996-152392205:- ENST00000420924 &lt;GRanges&gt;   </span>
<span class="co">#&gt; 10 chr2:179415981-179416357:- ENST00000342992 &lt;GRanges&gt;   </span>
<span class="co">#&gt; # … with 11 more rows</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="33-modify-transcripts-with-junctions">3.3 Modify transcripts with junctions<a class="anchor" aria-label="anchor" href="#33-modify-transcripts-with-junctions"></a>
</h3>
<p>We modify the canonical transcripts by introducing the splice junctions. Then we add the transcript sequence in a fixed-sized window around the junction positions, the context sequence.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">toy_junc_df</span>
<span class="co">#&gt; # A tibble: 14 × 2</span>
<span class="co">#&gt;    junc_id                    tx_id          </span>
<span class="co">#&gt;    &lt;chr&gt;                      &lt;chr&gt;          </span>
<span class="co">#&gt;  1 chr2:152389996-152392205:- ENST00000409198</span>
<span class="co">#&gt;  2 chr2:152389996-152390729:- ENST00000409198</span>
<span class="co">#&gt;  3 chr2:152389955-152389956:- ENST00000409198</span>
<span class="co">#&gt;  4 chr2:152388410-152392205:- ENST00000409198</span>
<span class="co">#&gt;  5 chr2:152388410-152390729:- ENST00000409198</span>
<span class="co">#&gt;  6 chr2:179415981-179416357:- ENST00000342992</span>
<span class="co">#&gt;  7 chr2:179415987-179415988:- ENST00000342992</span>
<span class="co">#&gt;  8 chr2:179415000-179416357:- ENST00000342992</span>
<span class="co">#&gt;  9 chr2:179445336-179446207:- ENST00000342992</span>
<span class="co">#&gt; 10 chr2:179446225-179446226:- ENST00000342992</span>
<span class="co">#&gt; 11 chr2:179445336-179446633:- ENST00000342992</span>
<span class="co">#&gt; 12 chr2:179642044-179642187:- ENST00000342992</span>
<span class="co">#&gt; 13 chr2:179642146-179642147:- ENST00000342992</span>
<span class="co">#&gt; 14 chr2:179642044-179642431:- ENST00000342992</span>


<span class="va">junc_df</span> <span class="op">&lt;-</span> <span class="va">toy_junc_df</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> 
  <span class="fu"><a href="reference/add_context_seq.html">add_context_seq</a></span><span class="op">(</span>transcripts <span class="op">=</span> <span class="va">toy_transcripts</span>, size <span class="op">=</span> <span class="fl">400</span>, bsg <span class="op">=</span> <span class="va">bsg</span><span class="op">)</span>


<span class="va">junc_df</span>
<span class="co">#&gt; # A tibble: 14 × 8</span>
<span class="co">#&gt;    junc_id      tx_id tx_mod_id junc_pos_tx cts_seq cts_junc_pos cts_size cts_id</span>
<span class="co">#&gt;    &lt;chr&gt;        &lt;chr&gt; &lt;chr&gt;           &lt;int&gt; &lt;chr&gt;          &lt;dbl&gt;    &lt;int&gt; &lt;chr&gt; </span>
<span class="co">#&gt;  1 chr2:152389… ENST… ENST0000…       16412 AAGAAG…          200      400 ef606…</span>
<span class="co">#&gt;  2 chr2:152389… ENST… ENST0000…       16517 AAGAAG…          200      400 6c189…</span>
<span class="co">#&gt;  3 chr2:152389… ENST… ENST0000…       17290 ACATCT…          200      400 c8bd5…</span>
<span class="co">#&gt;  4 chr2:152388… ENST… ENST0000…       16412 AAGAAG…          200      400 d41d2…</span>
<span class="co">#&gt;  5 chr2:152388… ENST… ENST0000…       16517 AAGAAG…          200      400 db9b3…</span>
<span class="co">#&gt;  6 chr2:179415… ENST… ENST0000…       83789 TGGATT…          200      400 744c1…</span>
<span class="co">#&gt;  7 chr2:179415… ENST… ENST0000…       84158 ATTTGA…          200      400 5315f…</span>
<span class="co">#&gt;  8 chr2:179415… ENST… ENST0000…       83789 TGGATT…          200      400 8eec0…</span>
<span class="co">#&gt;  9 chr2:179445… ENST… ENST0000…       59307 CGGGCT…          200      400 5ab65…</span>
<span class="co">#&gt; 10 chr2:179446… ENST… ENST0000…       59288 TTATCT…          200      400 c233b…</span>
<span class="co">#&gt; 11 chr2:179445… ENST… ENST0000…       58982 TGGCTA…          200      400 fddf5…</span>
<span class="co">#&gt; 12 chr2:179642… ENST… ENST0000…        4828 TAGAAG…          200      400 ce662…</span>
<span class="co">#&gt; 13 chr2:179642… ENST… ENST0000…        4868 TAGACC…          200      400 86af1…</span>
<span class="co">#&gt; 14 chr2:179642… ENST… ENST0000…        4703 GTCTCC…          200      400 ec963…</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="34-annotate-peptide-sequence">3.4 Annotate peptide sequence<a class="anchor" aria-label="anchor" href="#34-annotate-peptide-sequence"></a>
</h3>
<p>Finally, we use the splice junctions to modify the coding sequences (CDS) of the reference transcripts. The resulting CDS sequences are translated into protein sequence and further annotated with the peptide around the junction, the relative position of the splice junction in the peptide, and the location of the junction in an open reading frame (ORF).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">junc_df</span> <span class="op">&lt;-</span> <span class="va">junc_df</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> 
  <span class="fu"><a href="reference/add_peptide.html">add_peptide</a></span><span class="op">(</span>cds<span class="op">=</span><span class="va">toy_cds</span>, size <span class="op">=</span> <span class="fl">30</span>, bsg <span class="op">=</span> <span class="va">bsg</span><span class="op">)</span>

<span class="va">junc_df</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">junc_id</span>, <span class="va">junc_in_orf</span>, <span class="va">peptide_context</span>, <span class="va">peptide_context_junc_pos</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 14 × 4</span>
<span class="co">#&gt;    junc_id                    junc_in_orf peptide_context       peptide_context…</span>
<span class="co">#&gt;    &lt;chr&gt;                      &lt;lgl&gt;       &lt;chr&gt;                            &lt;dbl&gt;</span>
<span class="co">#&gt;  1 chr2:152389996-152392205:- TRUE        PINRHFKYATQLMNEIC                   14</span>
<span class="co">#&gt;  2 chr2:152389996-152390729:- TRUE        PRHLLAKTAGDQISQIC                   14</span>
<span class="co">#&gt;  3 chr2:152389955-152389956:- FALSE       &lt;NA&gt;                                NA</span>
<span class="co">#&gt;  4 chr2:152388410-152392205:- TRUE        PINRHFKYATQLMNEIKYRK…               14</span>
<span class="co">#&gt;  5 chr2:152388410-152390729:- TRUE        PRHLLAKTAGDQISQIKYRK…               14</span>
<span class="co">#&gt;  6 chr2:179415981-179416357:- TRUE        PSDPSKFTLAVSPVAGTPDY…               14</span>
<span class="co">#&gt;  7 chr2:179415987-179415988:- FALSE       &lt;NA&gt;                                NA</span>
<span class="co">#&gt;  8 chr2:179415000-179416357:- TRUE        PSDPSKFTLAVSPVVPPIVE…               14</span>
<span class="co">#&gt;  9 chr2:179445336-179446207:- TRUE        KHYPKDILSKYYQGDST                   14</span>
<span class="co">#&gt; 10 chr2:179446225-179446226:- TRUE        PSDVPDKHYPKDILSKYYQG…               14</span>
<span class="co">#&gt; 11 chr2:179445336-179446633:- TRUE        PSDASKAAYARDPQFPPEGE…               14</span>
<span class="co">#&gt; 12 chr2:179642044-179642187:- TRUE        TPSDSGEWTVVAQNRLWNIR                14</span>
<span class="co">#&gt; 13 chr2:179642146-179642147:- TRUE        RAGRSSISVILTVEGKMR                  14</span>
<span class="co">#&gt; 14 chr2:179642044-179642431:- TRUE        VVGRPMPETFWFHDAVEHQV…               14</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="4-dummy-workflow">4 Dummy workflow<a class="anchor" aria-label="anchor" href="#4-dummy-workflow"></a>
</h2>
<p>In the following a dummy example workflow how to integrate predict splicing effects from mutations or which detect expressed splice junctions from RNA-seq data to predict potential neoantigen candidates with splice2neo. A test case will be added later</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/TRON-Bioinformatics/splice2neo" class="external-link">splice2neo</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>
<span class="co"># load genome of choice</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg19</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/AnnotationDbi" class="external-link">AnnotationDbi</a></span><span class="op">)</span>

<span class="co"># this is an example of a transcript database (see below for details)</span>
<span class="va">gtf_url</span> <span class="op">&lt;-</span> <span class="st">"ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_34/GRCh37_mapping/gencode.v34lift37.annotation.gtf.gz"</span>
<span class="co"># parse GTF file as txdb object</span>
<span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu">GenomicFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html" class="external-link">makeTxDbFromGFF</a></span><span class="op">(</span><span class="va">gtf_url</span><span class="op">)</span>

<span class="va">transcripts</span> <span class="op">&lt;-</span>
  <span class="fu">GenomicFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcriptsBy.html" class="external-link">exonsBy</a></span><span class="op">(</span><span class="va">txdb</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tx"</span><span class="op">)</span>, use.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">transcripts_gr</span> <span class="op">&lt;-</span> <span class="fu">GenomicFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html" class="external-link">transcripts</a></span><span class="op">(</span><span class="va">txdb</span><span class="op">)</span>
<span class="co"># Build a GRangesList with cds composed of individual exon ranges</span>
<span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">GenomicFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcriptsBy.html" class="external-link">cdsBy</a></span><span class="op">(</span><span class="va">txdb</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tx"</span><span class="op">)</span>, use.name <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># canonical junctions</span>
<span class="co"># the user can choose the best suited data sets for canonical junctions</span>
<span class="co"># the object `canonical junction` should be a vector of canonical junctions in the junction format</span>
<span class="va">canonical_juncs</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1:33361245-33361511:-"</span>,
    <span class="st">"chr1:32849649-32852380:-"</span>,
    <span class="st">"chrom:start-end:strand"</span><span class="op">)</span>
<span class="co"># OPTIONAL: if canonical junctions are available as a bed file, the user can transform the bed file into the juntion format:</span>
<span class="co"># for bed files defined by exon-exon boundaries:</span>
<span class="va">canonical_juncs</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/bed_to_junc.html">bed_to_junc</a></span><span class="op">(</span>bed_file <span class="op">=</span> <span class="st">"/path/to/canonical/file.bed"</span>, type <span class="op">=</span> <span class="st">"exon-exon"</span><span class="op">)</span>
<span class="co"># for bed files defined by introns:</span>
<span class="va">canonical_juncs</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/bed_to_junc.html">bed_to_junc</a></span><span class="op">(</span>bed_file <span class="op">=</span> <span class="st">"/path/to/canonical/file.bed"</span>, type <span class="op">=</span> <span class="st">"intron"</span><span class="op">)</span>


<span class="co"># import RNA data</span>
<span class="va">dat_leafcutter</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="reference/leafcutter_transform.html">leafcutter_transform</a></span><span class="op">(</span>path <span class="op">=</span> <span class="st">"/your/path/to/leafcutter/results"</span><span class="op">)</span>
<span class="co"># supported events: exon_skip, intron_retention, alt_3prime, alt_5prime, mutex_exons</span>
<span class="va">dat_spladder</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="reference/spladder_transform.html">spladder_transform</a></span><span class="op">(</span>path <span class="op">=</span> <span class="st">"/your/path/to/spladder/results"</span><span class="op">)</span>
<span class="va">dat_rna</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="reference/generate_combined_dataset.html">generate_combined_dataset</a></span><span class="op">(</span>spladder_juncs <span class="op">=</span> <span class="va">dat_spladder</span>, leafcutter_juncs <span class="op">=</span> <span class="va">dat_spladder</span><span class="op">)</span>

<span class="co"># import &amp; transform SpliceAi results</span>
<span class="va">dat_spliceai</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="reference/parse_spliceai.html">parse_spliceai</a></span><span class="op">(</span>vcf_file <span class="op">=</span> <span class="st">"path/to/spliceai/file.vcf"</span><span class="op">)</span>
<span class="va">dat_splicai_formatted</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/format_spliceai.html">format_spliceai</a></span><span class="op">(</span><span class="va">dat_spliceai</span><span class="op">)</span>
<span class="va">dat_spliceai_annotated</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="reference/annotate_spliceai_junction.html">annotate_spliceai_junction</a></span><span class="op">(</span>var_df <span class="op">=</span> <span class="va">dat_splicai_formatted</span>,
                             transcripts <span class="op">=</span> <span class="va">transcripts</span>,
                             transcripts_gr <span class="op">=</span> <span class="va">transcripts_gr</span><span class="op">)</span>

<span class="co"># import &amp; transform MMSplice results</span>
<span class="va">dat_mmsplice</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/parse_mmsplice.html">parse_mmsplice</a></span><span class="op">(</span>infile <span class="op">=</span> <span class="st">"path/to/mmsplice/file.csv"</span><span class="op">)</span>
<span class="va">dat_mmsplice_annotated</span>  <span class="op">&lt;-</span>
  <span class="fu"><a href="reference/annotate_mmsplice.html">annotate_mmsplice</a></span><span class="op">(</span>mmsplice_df <span class="op">=</span> <span class="va">dat_mmsplice</span>, transcripts <span class="op">=</span> <span class="va">transcripts</span><span class="op">)</span>

<span class="co"># mutation-based junctions</span>
<span class="va">dat_mut</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="reference/combine_mut_junc.html">combine_mut_junc</a></span><span class="op">(</span>spliceai_juncs <span class="op">=</span> <span class="va">dat_spliceai_annotated</span>, mmsplice_juncs <span class="op">=</span> <span class="va">dat_mmsplice_annotated</span><span class="op">)</span>

<span class="co"># add information if junction is canonical and if found to be expressed by SplAdder or LeafCutter</span>
<span class="va">dat_mut</span> <span class="op">&lt;-</span> <span class="va">dat_mut</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu">mutate</span><span class="op">(</span>is_canonical <span class="op">=</span> <span class="fu"><a href="reference/is_canonical.html">is_canonical</a></span><span class="op">(</span><span class="va">junc_id</span>, ref_junc <span class="op">=</span> <span class="va">canonical_juncs</span>, exons_gr <span class="op">=</span> <span class="va">transcripts</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu">mutate</span><span class="op">(</span>is_in_rnaseq <span class="op">=</span> <span class="fu"><a href="reference/is_in_rnaseq.html">is_in_rnaseq</a></span><span class="op">(</span><span class="va">junc_id</span>, rna_juncs <span class="op">=</span> <span class="va">dat_rna</span><span class="op">$</span><span class="va">junc_id</span><span class="op">)</span><span class="op">)</span>

<span class="co"># remove canonical junctions for further downstream analysis</span>
<span class="va">dat_for_requantification</span> <span class="op">&lt;-</span> <span class="va">dat_mut</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">is_canonical</span><span class="op">)</span>

<span class="co"># add context sequences</span>
<span class="co"># a list of GRanges with the transcript needs to be added at the moment</span>
<span class="co"># this will be done within add_context_seq in a future version</span>
<span class="va">dat_for_requantification_cts</span> <span class="op">&lt;-</span> <span class="va">dat_for_requantification</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/add_context_seq.html">add_context_seq</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">400</span>, bsg <span class="op">=</span> <span class="va">BSgenome.Hsapiens.UCSC.hg19</span>, transcripts <span class="op">=</span> <span class="va">transcripts</span><span class="op">)</span>


<span class="co"># transform to easyquant-format</span>
<span class="va">dat_easyquant</span> <span class="op">&lt;-</span> <span class="va">dat_for_requantification_cts</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/transform_for_requant.html">transform_for_requant</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu">write_delim</span><span class="op">(</span><span class="va">dat_easyquant</span>, <span class="st">"path/to/easyquant/input/file.txt"</span>, delim <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>
<span class="co"># DO RE-QUANTIFICATION WITH EASYQUANT</span>
<span class="co"># https://github.com/TRON-Bioinformatics/easyquant</span>


<span class="co"># add peptide sequence</span>
<span class="va">dat_for_requantification_cts_peptide</span> <span class="op">&lt;-</span>
  <span class="va">dat_for_requantification_cts</span>  <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/add_peptide.html">add_peptide</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">30</span>, bsg <span class="op">=</span> <span class="va">BSgenome.Hsapiens.UCSC.hg19</span>, cds <span class="op">=</span> <span class="va">cds</span><span class="op">)</span>

<span class="co"># merge EasyQuant results with data</span>
<span class="va">dat_cts_peptide_requantification</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="reference/map_requant.html">map_requant</a></span><span class="op">(</span>path_to_easyquant_folder <span class="op">=</span> <span class="st">"/path/to/easyquant/output_folder"</span>,
              junc_tib <span class="op">=</span> <span class="va">dat_for_requantification_cts_peptide</span><span class="op">)</span>


<span class="co"># EasyQuant results can be imported without direct merging with data</span>
<span class="va">dat_requant</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="reference/read_requant.html">read_requant</a></span><span class="op">(</span>path_folder <span class="op">=</span> <span class="st">"/path/to/easyquant/output_folder"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="5-building-the-transcript-database">5 Building the transcript database<a class="anchor" aria-label="anchor" href="#5-building-the-transcript-database"></a>
</h2>
<p>A database of transcripts can be build from a GTF file, saved and re-loaded as follows:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># use gtf file of choice and transform into transcript database</span>
<span class="va">gtf_url</span> <span class="op">&lt;-</span> <span class="st">"ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_34/GRCh37_mapping/gencode.v34lift37.annotation.gtf.gz"</span>

<span class="co"># parse GTF file as txdb object</span>
<span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu">GenomicFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html" class="external-link">makeTxDbFromGFF</a></span><span class="op">(</span><span class="va">gtf_url</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">saveDb</a></span><span class="op">(</span><span class="va">txdb</span>, file <span class="op">=</span> <span class="st">"/path/to/transripts/txdb.sqlite"</span><span class="op">)</span>

<span class="co"># load </span>
<span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span><span class="st">"/path/to/transripts/txdb.sqlite"</span><span class="op">)</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/TRON-Bioinformatics/splice2neo/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing splice2neo</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Jonas Ibn-Salem <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-9886-113X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Franziska Lang <br><small class="roles"> Author </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/TRON-Bioinformatics/splice2neo/actions" class="external-link"><img src="https://github.com/TRON-Bioinformatics/splice2neo/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://codecov.io/gh/TRON-Bioinformatics/splice2neo?branch=master" class="external-link"><img src="https://codecov.io/gh/TRON-Bioinformatics/splice2neo/branch/master/graph/badge.svg" alt="Codecov test coverage"></a></li>
<li><a href="https://github.com/TRON-Bioinformatics/splice2neo" class="external-link"><img src="https://img.shields.io/badge/devel%20version-0.4.0-blue.svg"></a></li>
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="https://img.shields.io/badge/lifecycle-experimental-blue.svg"></a></li>
<li><a href="https://github.com/TRON-Bioinformatics/splice2neo/commits/dev" class="external-link"><img src="https://img.shields.io/github/last-commit/TRON-Bioinformatics/splice2neo.svg"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Jonas Ibn-Salem, Franziska Lang.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
