<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Aberrant splice junction analysis with associated mutations • splice2neo</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Aberrant splice junction analysis with associated mutations">
<meta property="og:description" content="This package provides functions for the analysis of alternative or 
 aberrant splicing junctions and their creation from or association with 
 somatic mutations. It integrates the output of several tools which predict 
 splicing effects from mutation or RNA-seq data into a common splice junction 
 format based on genomic coordinates. Splice junctions can be annotated with 
 affected transcript sequences, CDS, and resulting peptide sequences.  ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">splice2neo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/TRON-Bioinformatics/splice2neo/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="splice2neo" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#splice2neo" class="anchor"></a>splice2neo</h1></div>
<!-- badges: start -->

<p>This package provides functions for the analysis of alternative splicing junctions and their association with somatic mutations. It integrates the output of several tools which predict splicing effects from mutations or which detect expressed splice junctions from RNA-seq data into a standardized splice junction format based on genomic coordinates. Detected splice junctions can be filtered against canonical ones and annotated with affected transcript sequences, CDS, and resulting peptide sequences. The resulting tumor-specific splice junctions can encode neoantigens.</p>
<p>Website: <a href="https://tron-bioinformatics.github.io/splice2neo/" class="uri">https://tron-bioinformatics.github.io/splice2neo/</a></p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>This R package is not yet on <a href="https://CRAN.R-project.org">CRAN</a> or <a href="https://www.bioconductor.org/">Bioconductor</a>. Therefore, you need to install it from this repository.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## install.packages("remotes")</span>
<span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"TRON-Bioinformatics/splice2neo"</span><span class="op">)</span></code></pre></div>
</div>
<div id="example" class="section level2">
<h2 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h2>
<p>This is a basic example of how to use some functions.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/TRON-Bioinformatics/splice2neo">splice2neo</a></span><span class="op">)</span>

<span class="co"># load human genome reference sequence</span>
<span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">bsg</span> <span class="op">&lt;-</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg19</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html">BSgenome.Hsapiens.UCSC.hg19</a></span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: tidyverse</span>
<span class="co">#&gt; ── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──</span>
<span class="co">#&gt; ✓ ggplot2 3.3.3     ✓ purrr   0.3.4</span>
<span class="co">#&gt; ✓ tibble  3.1.2     ✓ dplyr   1.0.6</span>
<span class="co">#&gt; ✓ tidyr   1.1.3     ✓ stringr 1.4.0</span>
<span class="co">#&gt; ✓ readr   1.4.0     ✓ forcats 0.5.1</span>
<span class="co">#&gt; ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──</span>
<span class="co">#&gt; x dplyr::filter() masks stats::filter()</span>
<span class="co">#&gt; x dplyr::lag()    masks stats::lag()</span></code></pre></div>
<div id="example-data" class="section level3">
<h3 class="hasAnchor">
<a href="#example-data" class="anchor"></a>Example data</h3>
<p>We start with some example splice junctions provided with the package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">junc_df</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="op">(</span>
  junc_id <span class="op">=</span> <span class="va">toy_junc_id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">6</span>, <span class="fl">10</span><span class="op">)</span><span class="op">]</span>
<span class="op">)</span>

<span class="va">junc_df</span>
<span class="co">#&gt; # A tibble: 3 x 1</span>
<span class="co">#&gt;   junc_id                   </span>
<span class="co">#&gt;   &lt;chr&gt;                     </span>
<span class="co">#&gt; 1 chr2_152389996_152392205_-</span>
<span class="co">#&gt; 2 chr2_179415981_179416357_-</span>
<span class="co">#&gt; 3 chr2_179446225_179446226_-</span></code></pre></div>
</div>
<div id="add-transcripts" class="section level3">
<h3 class="hasAnchor">
<a href="#add-transcripts" class="anchor"></a>Add transcripts</h3>
<p>Next, we find the transcripts which are affected by the splice junctions.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">junc_df</span> <span class="op">&lt;-</span> <span class="va">junc_df</span> <span class="op"><a href="reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu"><a href="reference/add_tx.html">add_tx</a></span><span class="op">(</span><span class="va">toy_transcripts</span><span class="op">)</span>

<span class="va">junc_df</span>
<span class="co">#&gt; # A tibble: 21 x 3</span>
<span class="co">#&gt;    junc_id                    tx_id           tx_lst      </span>
<span class="co">#&gt;    &lt;chr&gt;                      &lt;chr&gt;           &lt;named list&gt;</span>
<span class="co">#&gt;  1 chr2_152389996_152392205_- ENST00000409198 &lt;GRanges&gt;   </span>
<span class="co">#&gt;  2 chr2_152389996_152392205_- ENST00000172853 &lt;GRanges&gt;   </span>
<span class="co">#&gt;  3 chr2_152389996_152392205_- ENST00000397345 &lt;GRanges&gt;   </span>
<span class="co">#&gt;  4 chr2_152389996_152392205_- ENST00000427231 &lt;GRanges&gt;   </span>
<span class="co">#&gt;  5 chr2_152389996_152392205_- ENST00000618972 &lt;GRanges&gt;   </span>
<span class="co">#&gt;  6 chr2_152389996_152392205_- ENST00000413693 &lt;GRanges&gt;   </span>
<span class="co">#&gt;  7 chr2_152389996_152392205_- ENST00000603639 &lt;GRanges&gt;   </span>
<span class="co">#&gt;  8 chr2_152389996_152392205_- ENST00000604864 &lt;GRanges&gt;   </span>
<span class="co">#&gt;  9 chr2_152389996_152392205_- ENST00000420924 &lt;GRanges&gt;   </span>
<span class="co">#&gt; 10 chr2_179415981_179416357_- ENST00000342992 &lt;GRanges&gt;   </span>
<span class="co">#&gt; # … with 11 more rows</span></code></pre></div>
</div>
<div id="modify-transcripts-with-junctions" class="section level3">
<h3 class="hasAnchor">
<a href="#modify-transcripts-with-junctions" class="anchor"></a>Modify transcripts with junctions</h3>
<p>We modify the canonical transcripts by introducing the splice junctions. Then we add the transcript sequence in a fixed-sized window around the junction positions, the context sequence.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">junc_df</span> <span class="op">&lt;-</span> <span class="va">junc_df</span> <span class="op"><a href="reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu"><a href="reference/add_context_seq.html">add_context_seq</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">400</span>, bsg <span class="op">=</span> <span class="va">bsg</span><span class="op">)</span>

<span class="va">junc_df</span> <span class="op"><a href="reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu">select</span><span class="op">(</span><span class="va">junc_id</span>, <span class="va">tx_id</span>, <span class="va">junc_pos_tx</span>, <span class="va">cts_seq</span>, <span class="va">cts_junc_pos</span>, <span class="va">cts_id</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 21 x 6</span>
<span class="co">#&gt;    junc_id     tx_id   junc_pos_tx cts_seq               cts_junc_pos cts_id    </span>
<span class="co">#&gt;    &lt;chr&gt;       &lt;chr&gt;         &lt;int&gt; &lt;chr&gt;                        &lt;dbl&gt; &lt;chr&gt;     </span>
<span class="co">#&gt;  1 chr2_15238… ENST00…       16412 AAGAAGACTTGACTTGGCTT…          199 ef6060403…</span>
<span class="co">#&gt;  2 chr2_15238… ENST00…       16412 AAGAAGACTTGACTTGGCTT…          199 ef6060403…</span>
<span class="co">#&gt;  3 chr2_15238… ENST00…       21515 AAGAAGACTTGACTTGGCTT…          199 729100c15…</span>
<span class="co">#&gt;  4 chr2_15238… ENST00…       21515 AAGAAGACTTGACTTGGCTT…          199 ef6060403…</span>
<span class="co">#&gt;  5 chr2_15238… ENST00…       21515 AAGAAGACTTGACTTGGCTT…          199 729100c15…</span>
<span class="co">#&gt;  6 chr2_15238… ENST00…        5502 AAGAAGACTTGACTTGGCTT…          199 ef6060403…</span>
<span class="co">#&gt;  7 chr2_15238… ENST00…       21312 AAGAAGACTTGACTTGGCTT…          199 729100c15…</span>
<span class="co">#&gt;  8 chr2_15238… ENST00…       21312 AAGAAGACTTGACTTGGCTT…          199 ef6060403…</span>
<span class="co">#&gt;  9 chr2_15238… ENST00…         576 AAGAAGACTTGACTTGGCTT…          199 8c2b828f5…</span>
<span class="co">#&gt; 10 chr2_17941… ENST00…       83789 TGGATTCCATGTTGAAAAGA…          199 744c11d66…</span>
<span class="co">#&gt; # … with 11 more rows</span></code></pre></div>
</div>
<div id="annotate-peptide-sequence" class="section level3">
<h3 class="hasAnchor">
<a href="#annotate-peptide-sequence" class="anchor"></a>Annotate peptide sequence</h3>
<p>Finally, we use the splice junctions to modify the coding sequences (CDS) of the reference transcripts. The resulting CDS sequences are translated into protein sequence and further annotated with the peptide around the junction, the relative position of the splice junction in the peptide, and the location of the junction in an open reading frame (ORF).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">junc_df</span> <span class="op">&lt;-</span> <span class="va">junc_df</span> <span class="op"><a href="reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu">mutate</span><span class="op">(</span>cds_lst <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">as.list</a></span><span class="op">(</span><span class="va">toy_cds</span><span class="op">[</span><span class="va">tx_id</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu"><a href="reference/add_peptide.html">add_peptide</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">30</span>, bsg <span class="op">=</span> <span class="va">bsg</span><span class="op">)</span>

<span class="va">junc_df</span> <span class="op"><a href="reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu">select</span><span class="op">(</span><span class="va">junc_id</span>, <span class="va">tx_id</span>, <span class="va">junc_in_orf</span>, <span class="va">peptide_context</span>, <span class="va">peptide_context_junc_pos</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 21 x 5</span>
<span class="co">#&gt;    junc_id        tx_id     junc_in_orf peptide_context       peptide_context_j…</span>
<span class="co">#&gt;    &lt;chr&gt;          &lt;chr&gt;     &lt;lgl&gt;       &lt;chr&gt;                              &lt;dbl&gt;</span>
<span class="co">#&gt;  1 chr2_15238999… ENST0000… TRUE        PINRHFKYATQLMNEIC                     14</span>
<span class="co">#&gt;  2 chr2_15238999… ENST0000… TRUE        PINRHFKYATQLMNEIC                     14</span>
<span class="co">#&gt;  3 chr2_15238999… ENST0000… TRUE        PINRHFKYATQLMNEIC                     14</span>
<span class="co">#&gt;  4 chr2_15238999… ENST0000… TRUE        PINRHFKYATQLMNEIC                     14</span>
<span class="co">#&gt;  5 chr2_15238999… ENST0000… TRUE        PINRHFKYATQLMNEIC                     14</span>
<span class="co">#&gt;  6 chr2_15238999… ENST0000… TRUE        PINRHFKYATQLMNEIC                     14</span>
<span class="co">#&gt;  7 chr2_15238999… ENST0000… TRUE        PINRHFKYATQLMNEIC                     14</span>
<span class="co">#&gt;  8 chr2_15238999… ENST0000… TRUE        PINRHFKYATQLMNEIC                     14</span>
<span class="co">#&gt;  9 chr2_15238999… ENST0000… TRUE        PINRHFKYATQLMNEIC                     14</span>
<span class="co">#&gt; 10 chr2_17941598… ENST0000… TRUE        PSDPSKFTLAVSPVAGTPDY…                 14</span>
<span class="co">#&gt; # … with 11 more rows</span></code></pre></div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/TRON-Bioinformatics/splice2neo/">https://​github.com/​TRON-Bioinformatics/​splice2neo/​</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Jonas Ibn-Salem <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-9886-113X" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Franziska Lang <br><small class="roles"> Author </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/TRON-Bioinformatics/splice2neo/actions"><img src="https://github.com/TRON-Bioinformatics/splice2neo/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://codecov.io/gh/TRON-Bioinformatics/splice2neo?branch=master"><img src="https://codecov.io/gh/TRON-Bioinformatics/splice2neo/branch/master/graph/badge.svg" alt="Codecov test coverage"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Jonas Ibn-Salem, Franziska Lang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
