---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# splice2neo

<!-- badges: start -->
[![R-CMD-check](https://github.com/TRON-Bioinformatics/splice2neo/workflows/R-CMD-check/badge.svg)](https://github.com/TRON-Bioinformatics/splice2neo/actions)
[![Codecov test coverage](https://codecov.io/gh/TRON-Bioinformatics/splice2neo/branch/master/graph/badge.svg)](https://codecov.io/gh/TRON-Bioinformatics/splice2neo?branch=master)
<!-- badges: end -->

This package provides functions for the analysis of alternative splicing
junctions and their association with somatic mutations. It integrates the output
of several tools which predict splicing effects from mutations or which detect
expressed splice junctions from RNA-seq data into a standardized splice junction
format based on genomic coordinates. Detected splice junctions can be filtered
against canonical ones and annotated with affected transcript sequences, CDS,
and resulting peptide sequences. The resulting tumor-specific splice junctions
can encode neoantigens.
 
Website: https://tron-bioinformatics.github.io/splice2neo/
 
## Installation

This R package is not yet on [CRAN](https://CRAN.R-project.org) or
[Bioconductor](https://www.bioconductor.org/). Therefore, you need to install it
from this repository. 

```{r, eval = FALSE}
## install.packages("remotes")
remotes::install_github("TRON-Bioinformatics/splice2neo")
```

## Example

This is a basic example of how to use some functions.

```{r}
library(splice2neo)

# load human genome reference sequence
requireNamespace("BSgenome.Hsapiens.UCSC.hg19", quietly = TRUE)
bsg <- BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19
require(tidyverse)

```

### Example data

We start with some example splice junctions provided with the package.
 
```{r}
junc_df <- tibble(
  junc_id = toy_junc_id[c(1, 6, 10)]
)

junc_df
```

### Add transcripts 

Next, we find the transcripts which are affected by the splice junctions.

```{r}
junc_df <- junc_df %>% 
  add_tx(toy_transcripts)

junc_df
```

### Modify transcripts with junctions

We modify the canonical transcripts by introducing the splice junctions. Then we
add the transcript sequence in a fixed-sized window around the junction
positions, the context sequence.


```{r}
junc_df <- junc_df %>% 
  add_context_seq(size = 400, bsg = bsg)

junc_df %>% 
  select(junc_id, tx_id, junc_pos_tx, cts_seq, cts_junc_pos, cts_id)

```

### Annotate peptide sequence

Finally, we use the splice junctions to modify the coding sequences (CDS) of the
reference transcripts. The resulting CDS sequences are translated into protein
sequence and further annotated with the peptide around the junction, the
relative position of the splice junction in the peptide, and the location of the
junction in an open reading frame (ORF).

```{r}

junc_df <- junc_df %>% 
  mutate(cds_lst = as.list(toy_cds[tx_id])) %>% 
  add_peptide(size = 30, bsg = bsg)

junc_df %>% 
  select(junc_id, tx_id, junc_in_orf, peptide_context, peptide_context_junc_pos)

```

