---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# splice2neo

<!-- badges: start -->
<!-- badges: end -->

This package provides functions for the analysis of alternative or 
 aberrant splicing junctions and their creation from or association with 
 somatic mutations. It integrates the output of several tools which predict 
 splicing effects from mutation or RNA-seq data into a common splice junction 
 format based on genomic coordinates. Splice junctions can be annotated with 
 affected transcript sequences, CDS, and resulting peptide sequences.  

Website: https://tron.pages.gitlab.rlp.net/splice2neo
 
## Installation

### Install this package from this GitLab

This R package is not yet on [CRAN](https://CRAN.R-project.org) or
[Bioconductor](https://www.bioconductor.org/). Therefore, you have to install it
form this GitLab repository. 

However, this repository is a private GitLab repository and therefore you have to 
create an personal access token (PAT) first. This is described [here](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html).
As Scope use `api` or `read_api`.  


``` r
install.packages("remotes") # if needed

remotes::install_gitlab("tron/splice2neo", host = "gitlab.rlp.net", auth_token = "YOUR_ACCESS_TOKEN")

```

## Example

This is a basic example which shows you how to solve a common problem:

```{r}
library(splice2neo)

# load human genome reference sequence
requireNamespace("BSgenome.Hsapiens.UCSC.hg19", quietly = TRUE)
bsg <- BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19
```

### Start with some example splice junctions

```{r}
# load toy example splice junctions
junc_df <- tibble::tibble(
  junc_id = toy_junc_id[c(1, 6, 10)]
)

junc_df
```

### add transcripts that are affected by the splice junction

```{r}
junc_df <- junc_df %>% 
  add_tx(toy_transcripts)

junc_df
```

### Modify transcripts by junctions and add context sequence around junction position

```{r}
junc_df <- junc_df %>% 
  add_context_seq(size = 400, bsg = bsg)

junc_df %>% 
  dplyr::select(junc_id, tx_id, junc_pos_tx, cts_seq, cts_junc_pos, cts_id)

```

### Add resulting CDS and pepetide sequence

```{r}

junc_df <- junc_df %>% 
  dplyr::mutate(cds_lst = as.list(toy_cds[tx_id])) %>% 
  add_peptide(size = 30, bsg = bsg)

junc_df %>% 
  dplyr::select(junc_id, tx_id, junc_in_orf, peptide_context, peptide_context_junc_pos)

```

