<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Annotates splice junctions with resulting CDS and peptide sequence. — add_peptide • splice2neo</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Annotates splice junctions with resulting CDS and peptide sequence. — add_peptide"><meta property="og:description" content="Annotates splice junctions with resulting CDS and peptide sequence."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">splice2neo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.6.14</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/splice2neo_workflow.html">Splice2neo example analysis</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/TRON-Bioinformatics/splice2neo/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Annotates splice junctions with resulting CDS and peptide sequence.</h1>
    <small class="dont-index">Source: <a href="https://github.com/TRON-Bioinformatics/splice2neo/blob/v0.6.14/R/add_peptide.R" class="external-link"><code>R/add_peptide.R</code></a></small>
    <div class="hidden name"><code>add_peptide.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Annotates splice junctions with resulting CDS and peptide sequence.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">add_peptide</span><span class="op">(</span><span class="va">df</span>, <span class="va">cds</span>, flanking_size <span class="op">=</span> <span class="fl">14</span>, bsg <span class="op">=</span> <span class="cn">NULL</span>, keep_ranges <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-df">df<a class="anchor" aria-label="anchor" href="#arg-df"></a></dt>
<dd><p>A data.frame with splice junctions in rows and at least the columns:</p><ul><li><p><code>junc_id</code> junction id consisting of genomic coordinates</p></li>
<li><p><code>tx_id</code> the ID of the affected transcript/CDS (see <code><a href="add_tx.html">add_tx</a></code>)</p></li>
</ul></dd>


<dt id="arg-cds">cds<a class="anchor" aria-label="anchor" href="#arg-cds"></a></dt>
<dd><p>as a named <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html" class="external-link">GRangesList</a></code> of coding sequences (CDS) ranges</p></dd>


<dt id="arg-flanking-size">flanking_size<a class="anchor" aria-label="anchor" href="#arg-flanking-size"></a></dt>
<dd><p>the number of wild-type amino acids flanking the junction or novel sequence caused by the splice junction to the left and to the right.
Frame shift junctions are flanked by wild-type amino acids to the left and are annotated until the first stop codon.</p></dd>


<dt id="arg-bsg">bsg<a class="anchor" aria-label="anchor" href="#arg-bsg"></a></dt>
<dd><p><code><a href="https://rdrr.io/pkg/BSgenome/man/BSgenome-class.html" class="external-link">BSgenome</a></code> object such as
<code><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html" class="external-link">BSgenome.Hsapiens.UCSC.hg19</a></code></p></dd>


<dt id="arg-keep-ranges">keep_ranges<a class="anchor" aria-label="anchor" href="#arg-keep-ranges"></a></dt>
<dd><p>Should GRanges of CDS and modified CDS be
kept? If TRUE, the list columns <code>cds_lst</code> and <code>cds_mod_lst</code> are added to the output.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A data.frame with the same rows as the input <code>df</code> but with the
following additional column(s):</p><ul><li><p><code>cds_mod_id</code> An identifier made from <code>tx_id</code> and <code>junc_id</code></p></li>
<li><p><code>junc_pos_cds</code> the junction position in the modified CDS sequence</p></li>
<li><p><code>frame_shift</code> Indicator whether junction leads to frame shift.</p></li>
<li><p><code>is_first_reading_frame</code> Indicator whether modified CDS sequence is
translated into <code>protein</code> sequence using the 1st reading frame
(i.e. reading is starting from the first nucleotide) for in-frame peptides.</p></li>
<li><p><code>normalized_cds_junc_pos</code> The normalized position of the junction in the
modified CDS sequence to the left junction side.</p></li>
<li><p><code>protein</code> The full protein sequence of the translated modified CDS.</p></li>
<li><p><code>normalized_protein_junc_pos</code> The normalized position of the junction in
the <code>protein</code> sequence to the left junction side.</p></li>
<li><p><code>peptide_context_junc_pos</code> The junction position relative to the <code>peptide_context</code> sequence</p></li>
<li><p><code>junc_in_orf</code> Indicator whether the junction is located in an open reading frame.</p></li>
<li><p><code>peptide_context_seq_raw</code> The peptide sequence around the junction including stop codons.</p></li>
<li><p><code>peptide_context</code> The peptide sequence around the junction truncated after stop codons.</p></li>
<li><p><code>truncated_cds</code> Indicator whether the mutated gene product is a truncated
from of the WT gene product. If TRUE, <code>peptide_context</code> = NA.</p></li>
<li><p><code>cds_description</code> Descriptor of of the mutated gene product. Can be one of
c("mutated cds", "truncated wt cds", "no mutated gene product", "no wt cds", "not in ORF")</p></li>
</ul><p>If the <code>keep_ranges</code> is TRUE, the following additional columns are added to
the output data.frame:</p><ul><li><p><code>cds_lst</code> A list of <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> with
the original CDS as provided in <code>tx_id</code> column and <code>cds</code> object..</p></li>
<li><p><code>cds_mod_lst</code> A list of <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> with the
modified CDS ranges.</p></li>
</ul></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bsg</span> <span class="op">&lt;-</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg19</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html" class="external-link">BSgenome.Hsapiens.UCSC.hg19</a></span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">add_peptide</span><span class="op">(</span><span class="va">toy_junc_df</span>, <span class="va">toy_cds</span>, bsg <span class="op">=</span> <span class="va">bsg</span>, flanking_size <span class="op">=</span> <span class="fl">13</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 17 × 18</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    junc_id          tx_id protein protein_wt frame_shift cds_mod_id junc_pos_cds</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> chr2:152389996-… ENST… MADDED… MADDEDYEE… TRUE        ENST00000…        <span style="text-decoration: underline;">16</span>209</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> chr2:152389996-… ENST… MADDED… MADDEDYEE… TRUE        ENST00000…        <span style="text-decoration: underline;">16</span>314</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> chr2:152389955-… ENST… MADDED… MADDEDYEE… FALSE       ENST00000…        <span style="text-decoration: underline;">21</span>417</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> chr2:152388410-… ENST… MADDED… MADDEDYEE… FALSE       ENST00000…        <span style="text-decoration: underline;">16</span>209</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> chr2:152388410-… ENST… MADDED… MADDEDYEE… FALSE       ENST00000…        <span style="text-decoration: underline;">16</span>314</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> chr2:179415981-… ENST… MTTQAP… MTTQAPTFT… FALSE       ENST00000…        <span style="text-decoration: underline;">83</span>566</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> chr2:179415987-… ENST… MTTQAP… MTTQAPTFT… FALSE       ENST00000…        <span style="text-decoration: underline;">83</span>935</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> chr2:179415000-… ENST… MTTQAP… MTTQAPTFT… FALSE       ENST00000…        <span style="text-decoration: underline;">83</span>566</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> chr2:179445336-… ENST… MTTQAP… MTTQAPTFT… TRUE        ENST00000…        <span style="text-decoration: underline;">59</span>084</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> chr2:179446225-… ENST… MTTQAP… MTTQAPTFT… TRUE        ENST00000…        <span style="text-decoration: underline;">59</span>065</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">11</span> chr2:179445336-… ENST… MTTQAP… MTTQAPTFT… FALSE       ENST00000…        <span style="text-decoration: underline;">58</span>759</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">12</span> chr2:179642044-… ENST… MTTQAP… MTTQAPTFT… TRUE        ENST00000…         <span style="text-decoration: underline;">4</span>605</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">13</span> chr2:179642146-… ENST… MTTQAP… MTTQAPTFT… FALSE       ENST00000…         <span style="text-decoration: underline;">4</span>645</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">14</span> chr2:179642044-… ENST… MTTQAP… MTTQAPTFT… FALSE       ENST00000…         <span style="text-decoration: underline;">4</span>480</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">15</span> chr2:152226533-… ENST… <span style="color: #BB0000;">NA</span>      <span style="color: #BB0000;">NA</span>         <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span>                   <span style="color: #BB0000;">NA</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">16</span> chr2:152222731-… ENST… <span style="color: #BB0000;">NA</span>      <span style="color: #BB0000;">NA</span>         <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span>                   <span style="color: #BB0000;">NA</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">17</span> chr2:152388410-… ENST… MADDED… MADDEDYEE… FALSE       ENST00000…        <span style="text-decoration: underline;">22</span>962</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 11 more variables: protein_junc_pos &lt;dbl&gt;, normalized_cds_junc_pos &lt;int&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   normalized_protein_junc_pos_pre &lt;dbl&gt;, is_first_reading_frame &lt;lgl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   normalized_protein_junc_pos &lt;dbl&gt;, junc_in_orf &lt;lgl&gt;, truncated_cds &lt;lgl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   cds_description &lt;chr&gt;, peptide_context_seq_raw &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   peptide_context_junc_pos &lt;dbl&gt;, peptide_context &lt;chr&gt;</span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jonas Ibn-Salem, Franziska Lang, Michelle Cervante, Johannes Hausmann, Patrick Sorn.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

