<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ splice2neo</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">splice2neo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/TRON-Bioinformatics/splice2neo/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-parse-spliceai" class="hasAnchor"><a href="#section-parse-spliceai" class="anchor"></a>Parse SpliceAI</h2>
          <p class="section-desc"><p>Functions to parse SpliceAI output into standard splice junction format</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="parse_spliceai.html">parse_spliceai()</a></code> </p>
        </td>
        <td><p>Parse VCF output file from spliceAI as table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="format_spliceai.html">format_spliceai()</a></code> </p>
        </td>
        <td><p>Formats spliceAI output and filter for predicted effects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="annotate_spliceai_junction.html">annotate_spliceai_junction()</a></code> </p>
        </td>
        <td><p>Annotate splice variants with resulting junctions</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-parse-mmsplice" class="hasAnchor"><a href="#section-parse-mmsplice" class="anchor"></a>Parse MMsplice</h2>
          <p class="section-desc"><p>Functions to parse MMsplice output into standard splice junction format</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="parse_mmsplice.html">parse_mmsplice()</a></code> </p>
        </td>
        <td><p>Parse .csv file from MMsplice output as data.frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="annotate_mmsplice.html">annotate_mmsplice()</a></code> </p>
        </td>
        <td><p>Annotates the mmsplice output with additional columns including the junction as junc_id.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_exon_inclusion_junction.html">get_exon_inclusion_junction()</a></code> </p>
        </td>
        <td><p>Compute the resulting junctions (junc_id) for exon inclusion of given exon
and transcript.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_exon_skipping_junction.html">get_exon_skipping_junction()</a></code> </p>
        </td>
        <td><p>Compute the resulting junction (junc_id) for exon skipping of given exon
and transcript.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-parse-leafcutter" class="hasAnchor"><a href="#section-parse-leafcutter" class="anchor"></a>Parse LeafCutter</h2>
          <p class="section-desc"><p>Functions to parse LeafCutter output into standard splice junction format</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="import_leafcutter_bam.html">import_leafcutter_bam()</a></code> </p>
        </td>
        <td><p>Imports "_Aligned.out.bam.junc" file from Leafcutter output</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="import_leafcutter_counts.html">import_leafcutter_counts()</a></code> </p>
        </td>
        <td><p>Imports "_perind.counts.gz" from Leafcutter output.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="leafcutter_transform.html">leafcutter_transform()</a></code> </p>
        </td>
        <td><p>Imports "_perind.counts.gz" and "_Aligned.out.bam.junc" from Leafcutter output and transforms the raw output
into standardized junction output format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="leafcutter_transform_format.html">leafcutter_transform_format()</a></code> </p>
        </td>
        <td><p>Transforms Leafcutter intermediate files into standardized format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="transform_leafcutter_bam.html">transform_leafcutter_bam()</a></code> </p>
        </td>
        <td><p>Transforms Leafcutter bam junc file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="transform_leafcutter_counts.html">transform_leafcutter_counts()</a></code> </p>
        </td>
        <td><p>Transforms Leafcutter counts file into standardized junction format.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-parse-spladder" class="hasAnchor"><a href="#section-parse-spladder" class="anchor"></a>Parse SplAdder</h2>
          <p class="section-desc"><p>Functions to parse SplAdder output into standard splice junction format</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="import_spladder.html">import_spladder()</a></code> </p>
        </td>
        <td><p>Imports SPLADDER output from a given path with ".confirmed.txt.gz" files.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spladder_transform.html">spladder_transform()</a></code> </p>
        </td>
        <td><p>Imports SPLADDER output from a given path and transforms it into standardized
junction format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spladder_transform_format.html">spladder_transform_format()</a></code> </p>
        </td>
        <td><p>Transforms Spladder output into standardized junction format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spladder_transform_a3ss.html">spladder_transform_a3ss()</a></code> </p>
        </td>
        <td><p>Transforms events from alternative 3' splice sites from SPLADDER output
format into standardized junction format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spladder_transform_a5ss.html">spladder_transform_a5ss()</a></code> </p>
        </td>
        <td><p>Transforms events from alternative 5' splice sites from SPLADDER output
format into standardized junction format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spladder_transform_intron_retention.html">spladder_transform_intron_retention()</a></code> </p>
        </td>
        <td><p>Transforms events resulting from intron retention from SPLADDER output
format into standardized junction format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spladder_transform_mutex_exon.html">spladder_transform_mutex_exon()</a></code> </p>
        </td>
        <td><p>Transforms events resulting from mutually exclusive exons from SPLADDER output
format into standardized junction format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spladder_transform_exon_skipping.html">spladder_transform_exon_skipping()</a></code> </p>
        </td>
        <td><p>Transforms events resulting from exon skipping from SPLADDER output
format into standardized junction format</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-annotate" class="hasAnchor"><a href="#section-annotate" class="anchor"></a>Annotate</h2>
          <p class="section-desc"><p>Functions to annotate splice junctions</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="add_tx.html">add_tx()</a></code> </p>
        </td>
        <td><p>Annotate splice junctions with affected transcript IDs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="add_context_seq.html">add_context_seq()</a></code> </p>
        </td>
        <td><p>Annotate splice junctions with resulting transcript sequence</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="modify_tx.html">modify_tx()</a></code> </p>
        </td>
        <td><p>modifies transcripts by applying splice-junctions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_junc_pos.html">get_junc_pos()</a></code> </p>
        </td>
        <td><p>Get the position of input junction in the transcript sequences</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="add_peptide.html">add_peptide()</a></code> </p>
        </td>
        <td><p>Annotate splice junctions with resulting CDS and peptide sequence</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seq_truncate_nonstop.html">seq_truncate_nonstop()</a></code> </p>
        </td>
        <td><p>Truncate input sequence after input position before next stop codons (<code>*</code>).</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-requantification" class="hasAnchor"><a href="#section-requantification" class="anchor"></a>Requantification</h2>
          <p class="section-desc"><p>Functions for read support re-quantification with easyquant</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="map_requant.html">map_requant()</a></code> </p>
        </td>
        <td><p>Maps the re-quantification result from Easyquant on the junction-transcript centric tibble by hash id.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_requant.html">read_requant()</a></code> </p>
        </td>
        <td><p>Imports the re-quantification results from analysis with easyquant.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="transform_for_requant.html">transform_for_requant()</a></code> </p>
        </td>
        <td><p>Creates a table with context sequences for re-quantification analysis with Easyquant.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-canonical" class="hasAnchor"><a href="#section-canonical" class="anchor"></a>Canonical</h2>
          <p class="section-desc"><p>Functions annotate and filter for cnaonical splicing</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="is_canonical.html">is_canonical()</a></code> </p>
        </td>
        <td><p>test if junction is canonical junction</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-merge-and-filter" class="hasAnchor"><a href="#section-merge-and-filter" class="anchor"></a>Merge and filter</h2>
          <p class="section-desc"><p>Functions annotate and filter for cnaonical splicing</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="is_in_rnaseq.html">is_in_rnaseq()</a></code> </p>
        </td>
        <td><p>test if junction was found in corresponding RNA-seq data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="combine_mut_junc.html">combine_mut_junc()</a></code> </p>
        </td>
        <td><p>Combines tibbles with junctions from SpliceAI and MMsplice</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="add_identified_in_RNA.html">add_identified_in_RNA()</a></code> </p>
        </td>
        <td><p>This is a wrapper function to directly map the information if a junction predicted from WES data was found in RNA-seq by leafcutter or spladder</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="generate_combined_dataset.html">generate_combined_dataset()</a></code> </p>
        </td>
        <td><p>Combines tibbles with junctions from Spladder and Leafcutter into a combined dataset of expressed splice junctions</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-others" class="hasAnchor"><a href="#section-others" class="anchor"></a>Others</h2>
          <p class="section-desc"><p>Other functions</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="junc_to_gr.html">junc_to_gr()</a></code> </p>
        </td>
        <td><p>Convert a splice junction ID into a GRanges object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="change_to_junction_rules.html">change_to_junction_rules</a></code> </p>
        </td>
        <td><p>Rules on how a splicing affecting variant creates a junction</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="next_junctions.html">next_junctions()</a></code> </p>
        </td>
        <td><p>Get genomic coordinates of possible next donor and acceptor sides.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sort_columns.html">sort_columns()</a></code> </p>
        </td>
        <td><p>Sorts columns of junction output file in the following order:
"junction_start", "junction_end", "strand", "chromosome", "Gene",
"class", "AS_event_ID", "junction_id"</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="unsorted_junc_df.html">unsorted_junc_df</a></code> </p>
        </td>
        <td><p>A tibble in junction format but unsorted columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="transform_leafcutter_bam.html">transform_leafcutter_bam()</a></code> </p>
        </td>
        <td><p>Transforms Leafcutter bam junc file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="transform_leafcutter_counts.html">transform_leafcutter_counts()</a></code> </p>
        </td>
        <td><p>Transforms Leafcutter counts file into standardized junction format.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-data" class="hasAnchor"><a href="#section-data" class="anchor"></a>Data</h2>
          <p class="section-desc"><p>Toy example datasets</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="toy_cds.html">toy_cds</a></code> </p>
        </td>
        <td><p>An example CDS annotation GRangesList from a subset of human genome</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="toy_junc_df.html">toy_junc_df</a></code> </p>
        </td>
        <td><p>An example dataset of 18 splice junctions from human
(hg19) genome as a data.frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="toy_junc_id.html">toy_junc_id</a></code> </p>
        </td>
        <td><p>An example dataset of 18 splice junctions in <code>junc_id</code> fomrat from human
(hg19) genome as character vector</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="toy_junc_id_enst.html">toy_junc_id_enst</a></code> </p>
        </td>
        <td><p>An example dataset of 18 transcript IDs matching to the junctions in <code>toy_junc_id</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="toy_transcripts.html">toy_transcripts</a></code> </p>
        </td>
        <td><p>An example transcript annotation GRangesList from a subset of human genome</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="toy_transcripts_gr.html">toy_transcripts_gr</a></code> </p>
        </td>
        <td><p>An example dataset of full transcript ranges from a subset of human genome.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="unsorted_junc_df.html">unsorted_junc_df</a></code> </p>
        </td>
        <td><p>A tibble in junction format but unsorted columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spladder_output.a3ss.html">spladder_output.a3ss</a></code> </p>
        </td>
        <td><p>A tibble containing events from alternative 3' splice sites identified by
Spladder</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spladder_output.a5ss.html">spladder_output.a5ss</a></code> </p>
        </td>
        <td><p>A tibble containing events from alternative 5' splice sites identified by
Spladder</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spladder_output.exonskip.html">spladder_output.exonskip</a></code> </p>
        </td>
        <td><p>A tibble containing events from exon skipping identified by
Spladder</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spladder_output.intronreten.html">spladder_output.intronreten</a></code> </p>
        </td>
        <td><p>A tibble containing events from intron retention identified by
Spladder</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spladder_output.mutexon.html">spladder_output.mutexon</a></code> </p>
        </td>
        <td><p>A tibble containing events from mutually exclusive exons identified by
Spladder</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spladder_output.html">spladder_output</a></code> </p>
        </td>
        <td><p>A list of tibbles with each list containing events from a different
splicing class identified by Spladder</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="leafcutter_juncs.html">leafcutter_juncs</a></code> </p>
        </td>
        <td><p>A tibble with junction in standardized format. example data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spladder_juncs.html">spladder_juncs</a></code> </p>
        </td>
        <td><p>A tibble with junction in standardized format.  example data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="canonical_juncs.html">canonical_juncs</a></code> </p>
        </td>
        <td><p>A tibble with canonical junctions and their source (comma, separated).
example data</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Jonas Ibn-Salem, Franziska Lang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


