<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Package index â€¢ splice2neo</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">splice2neo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.6.13</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/splice2neo_workflow.html">Splice2neo example analysis</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/TRON-Bioinformatics/splice2neo/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="parse-spliceai">Parse SpliceAI <a href="#parse-spliceai" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to parse SpliceAI and CI-SpliceAI output into standard splice junction format</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="parse_spliceai.html">parse_spliceai()</a></code> </p>
        </td>
        <td><p>Parse VCF output file from spliceAI as table</p></td>
      </tr><tr><td>
          <p><code><a href="parse_spliceai_thresh.html">parse_spliceai_thresh()</a></code> </p>
        </td>
        <td><p>Parse VCF output file from SpliceAI with -T flag as table</p></td>
      </tr><tr><td>
          <p><code><a href="parse_cispliceai_thresh.html">parse_cispliceai_thresh()</a></code> </p>
        </td>
        <td><p>Parse VCF output file from CI-SpliceAI with -t flag as table</p></td>
      </tr><tr><td>
          <p><code><a href="format_spliceai.html">format_spliceai()</a></code> </p>
        </td>
        <td><p>Formats spliceAI output and filter for predicted effects</p></td>
      </tr><tr><td>
          <p><code><a href="format_spliceai_thresh.html">format_spliceai_thresh()</a></code> </p>
        </td>
        <td><p>Formats SpliceAI with -T flag output and filter for predicted effects</p></td>
      </tr><tr><td>
          <p><code><a href="format_cispliceai_thresh.html">format_cispliceai_thresh()</a></code> </p>
        </td>
        <td><p>Formats CI-SpliceAI with -t flag output and filter for predicted effects</p></td>
      </tr><tr><td>
          <p><code><a href="annotate_mut_effect.html">annotate_mut_effect()</a></code> </p>
        </td>
        <td><p>Annotate splice variants effects with resulting junctions</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="parse-mmsplice">Parse MMsplice <a href="#parse-mmsplice" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to parse MMsplice output into standard splice junction format</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="parse_mmsplice.html">parse_mmsplice()</a></code> </p>
        </td>
        <td><p>Parse .csv file from MMsplice output as data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="annotate_mmsplice.html">annotate_mmsplice()</a></code> </p>
        </td>
        <td><p>Annotates the mmsplice output with additional columns including the junction as junc_id.</p></td>
      </tr><tr><td>
          <p><code><a href="get_exon_inclusion_junction.html">get_exon_inclusion_junction()</a></code> </p>
        </td>
        <td><p>Compute the resulting junctions (junc_id) for exon inclusion of given exon and transcript.</p></td>
      </tr><tr><td>
          <p><code><a href="get_exon_skipping_junction.html">get_exon_skipping_junction()</a></code> </p>
        </td>
        <td><p>Compute the resulting junction (junc_id) for exon skipping of given exon and transcript.</p></td>
      </tr><tr><td>
          <p><code><a href="unique_junc_mmsplice.html">unique_junc_mmsplice()</a></code> </p>
        </td>
        <td><p>Filter junction datasets for unique <code>mut_id</code>, <code>junc_id</code>, <code>tx_id</code>. If there were multiple exon inclusion events predicted to lead to the same combination of <code>mut_id</code>, <code>junc_id</code>, <code>tx_id</code>, only the effect with the maximal score is returned.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="parse-pangolin">Parse Pangolin <a href="#parse-pangolin" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to parse Pangolin output into standard splice junction format</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="parse_pangolin.html">parse_pangolin()</a></code> </p>
        </td>
        <td><p>Parse VCF output file from pangolin as table</p></td>
      </tr><tr><td>
          <p><code><a href="format_pangolin.html">format_pangolin()</a></code> </p>
        </td>
        <td><p>Formats pangolin output and filter for predicted effects</p></td>
      </tr><tr><td>
          <p><code><a href="annotate_mut_effect.html">annotate_mut_effect()</a></code> </p>
        </td>
        <td><p>Annotate splice variants effects with resulting junctions</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="parse-leafcutter">Parse LeafCutter <a href="#parse-leafcutter" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to parse LeafCutter output into standard splice junction format</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="import_leafcutter_counts.html">import_leafcutter_counts()</a></code> </p>
        </td>
        <td><p>Imports "_perind.counts.gz" from LeafCutter output.</p></td>
      </tr><tr><td>
          <p><code><a href="leafcutter_transform.html">leafcutter_transform()</a></code> </p>
        </td>
        <td><p>Imports "*perind.counts.gz" from LeafCutter output and transforms the raw output into standardized junction output format. Results of one patient should be stored in the given path.</p></td>
      </tr><tr><td>
          <p><code><a href="transform_leafcutter_counts.html">transform_leafcutter_counts()</a></code> </p>
        </td>
        <td><p>Transforms LeafCutter counts file into standardized junction format.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="parse-regtools">Parse Regtools <a href="#parse-regtools" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to parse Regtools output into standard splice junction format</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="import_regtools_junc.html">import_regtools_junc()</a></code> </p>
        </td>
        <td><p>Imports Regtools junctions annotate table</p></td>
      </tr><tr><td>
          <p><code><a href="regtools_transform.html">regtools_transform()</a></code> </p>
        </td>
        <td><p>Imports Regtools junctions annotate output and transforms the raw output into standardized junction output format</p></td>
      </tr><tr><td>
          <p><code><a href="transform_regtools_junc.html">transform_regtools_junc()</a></code> </p>
        </td>
        <td><p>Transforms Regtools intermediate files into standardized format</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="parse-spladder">Parse SplAdder <a href="#parse-spladder" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to parse SplAdder output into standard splice junction format</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="import_spladder.html">import_spladder()</a></code> </p>
        </td>
        <td><p>Imports SplAdder output from a given path with ".confirmed.txt.gz" files. The results for one patient should be stored in the given path. Please note that multiple (coordinated) exons skips (mult_exon_skip) are currently not supported.</p></td>
      </tr><tr><td>
          <p><code><a href="spladder_transform.html">spladder_transform()</a></code> </p>
        </td>
        <td><p>Imports SplAdder output from a given path and transforms it into standardized junction format. Results from one patients should be stored per folder.</p></td>
      </tr><tr><td>
          <p><code><a href="spladder_transform_format.html">spladder_transform_format()</a></code> </p>
        </td>
        <td><p>Transforms SplAdder output into standardized junction format</p></td>
      </tr><tr><td>
          <p><code><a href="spladder_transform_ass.html">spladder_transform_ass()</a></code> </p>
        </td>
        <td><p>Transforms events from alternative 3' or 5' splice sites from SplAdder output format into standardized junction format</p></td>
      </tr><tr><td>
          <p><code><a href="spladder_transform_intron_retention.html">spladder_transform_intron_retention()</a></code> </p>
        </td>
        <td><p>Transforms events resulting from intron retention from SplAdder output format into standardized junction format</p></td>
      </tr><tr><td>
          <p><code><a href="spladder_transform_mutex_exon.html">spladder_transform_mutex_exon()</a></code> </p>
        </td>
        <td><p>Transforms events resulting from mutually exclusive exons from SplAdder output format into standardized junction format</p></td>
      </tr><tr><td>
          <p><code><a href="spladder_transform_exon_skipping.html">spladder_transform_exon_skipping()</a></code> </p>
        </td>
        <td><p>Transforms events resulting from exon skipping from SplAdder output format into standardized junction format</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="parse-irfinder">Parse IRfinder <a href="#parse-irfinder" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to parse IRfinder output into standard splice junction format</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="import_irfinder_txt.html">import_irfinder_txt()</a></code> </p>
        </td>
        <td><p>Imports tabular IRFinder retained introns predictions</p></td>
      </tr><tr><td>
          <p><code><a href="filter_irfinder_txt.html">filter_irfinder_txt()</a></code> </p>
        </td>
        <td><p>Filter IRFinder intermediate table to remove likely false positive IR predictions</p></td>
      </tr><tr><td>
          <p><code><a href="parse_irfinder_txt.html">parse_irfinder_txt()</a></code> </p>
        </td>
        <td><p>Imports "IRFinder-IR-nondir.txt" from IRFinder short mode and transforms the raw output into standardized junction output format.</p></td>
      </tr><tr><td>
          <p><code><a href="transform_irfinder_txt.html">transform_irfinder_txt()</a></code> </p>
        </td>
        <td><p>Transforms IRFinder intermediate table into standardized junction format</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="parse-suppa-">Parse SUPPA2 <a href="#parse-suppa-" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to parse SUPPA2 output into standard splice junction format</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="read_suppa_ioe.html">read_suppa_ioe()</a></code> </p>
        </td>
        <td><p>Imports "_strict.ioe" from SUPPA2 output.</p></td>
      </tr><tr><td>
          <p><code><a href="transform_suppa_se_events.html">transform_suppa_se_events()</a></code> </p>
        </td>
        <td><p>Transforms events resulting from exon skipping from SUPPA2 output format into standardized junction format</p></td>
      </tr><tr><td>
          <p><code><a href="transform_suppa_ir_events.html">transform_suppa_ir_events()</a></code> </p>
        </td>
        <td><p>Transforms events resulting from intron retention from SUPPA2 output format into standardized junction format</p></td>
      </tr><tr><td>
          <p><code><a href="transform_suppa_ass_events.html">transform_suppa_ass_events()</a></code> </p>
        </td>
        <td><p>Transforms events from alternative 3' or 5' splice sites from SUPPA2 output format into standardized junction format</p></td>
      </tr><tr><td>
          <p><code><a href="transform_suppa_mxe_events.html">transform_suppa_mxe_events()</a></code> </p>
        </td>
        <td><p>Transforms events resulting from mutually exclusive exons from SUPPA2 output format into standardized junction format</p></td>
      </tr><tr><td>
          <p><code><a href="suppa_transform_format.html">suppa_transform_format()</a></code> </p>
        </td>
        <td><p>Transforms SUPPA2 ioe event files file into standardized junction format.</p></td>
      </tr><tr><td>
          <p><code><a href="suppa_import.html">suppa_import()</a></code> </p>
        </td>
        <td><p>Imports SUPPA2 output from a given path with "_strict.ioe" files. The results for one patient should be stored in the given path. Please note that AF and AL events are not supported.</p></td>
      </tr><tr><td>
          <p><code><a href="suppa_transform.html">suppa_transform()</a></code> </p>
        </td>
        <td><p>Imports SUPPA2 output from a given path and transforms it into standardized junction format. Results from one patients should be stored per folder. Please note that alternative first exons (AF) and alternative last exons (AL) are not supported.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="parse-star">Parse STAR <a href="#parse-star" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to parse STAR SJ.out.tab into standard splice junction format</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="parse_star_sj.html">parse_star_sj()</a></code> </p>
        </td>
        <td><p>Imports "*SJ.out.tab" from STAR and transforms the raw output into standardized junction output format.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="parse-stringtie">Parse StringTie <a href="#parse-stringtie" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to parse StringTie into standard splice junction format</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="import_stringtie_gtf.html">import_stringtie_gtf()</a></code> </p>
        </td>
        <td><p>Imports StringTie GTF file</p></td>
      </tr><tr><td>
          <p><code><a href="stringtie_transform_format.html">stringtie_transform_format()</a></code> </p>
        </td>
        <td><p>Transforms StringTie intermediate table into standardized junction format</p></td>
      </tr><tr><td>
          <p><code><a href="stringtie_transform.html">stringtie_transform()</a></code> </p>
        </td>
        <td><p>Imports StringTie assembled transcripts and transforms the raw output into standardized junction output format</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="parse-other-formats">Parse other formats <a href="#parse-other-formats" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to parse otherformats into standard splice junction format</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="bed_to_junc.html">bed_to_junc()</a></code> </p>
        </td>
        <td><p>Transform a bed file into junction format</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="annotate">Annotate <a href="#annotate" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to annotate splice junctions</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="add_tx.html">add_tx()</a></code> </p>
        </td>
        <td><p>Annotate splice junctions with all possible transcript IDs in the given genomic region.</p></td>
      </tr><tr><td>
          <p><code><a href="add_context_seq.html">add_context_seq()</a></code> </p>
        </td>
        <td><p>Annotate splice junctions with resulting transcript sequence</p></td>
      </tr><tr><td>
          <p><code><a href="modify_tx.html">modify_tx()</a></code> </p>
        </td>
        <td><p>Modify transcript by introducing splice junctions</p></td>
      </tr><tr><td>
          <p><code><a href="get_junc_pos.html">get_junc_pos()</a></code> </p>
        </td>
        <td><p>Get the position of input junction in the transcript sequences</p></td>
      </tr><tr><td>
          <p><code><a href="get_intronretention_genomic_alt_pos.html">get_intronretention_genomic_alt_pos()</a></code> </p>
        </td>
        <td><p>Get the alternative position of input junction from an intron retention event in the genomic sequences</p></td>
      </tr><tr><td>
          <p><code><a href="get_unknown_exon_intron_retention.html">get_unknown_exon_intron_retention()</a></code> </p>
        </td>
        <td><p>Get the exon range flanking the other side of an intron retention of interest</p></td>
      </tr><tr><td>
          <p><code><a href="add_peptide.html">add_peptide()</a></code> </p>
        </td>
        <td><p>Annotates splice junctions with resulting CDS and peptide sequence.</p></td>
      </tr><tr><td>
          <p><code><a href="seq_truncate_nonstop.html">seq_truncate_nonstop()</a></code> </p>
        </td>
        <td><p>Truncate input sequence after input position before next stop codons (<code>*</code>).</p></td>
      </tr><tr><td>
          <p><code><a href="exon_in_intron.html">exon_in_intron()</a></code> </p>
        </td>
        <td><p>Annotate if there is an exon within an intron</p></td>
      </tr><tr><td>
          <p><code><a href="is_start_on_exon.html">is_start_on_exon()</a></code> </p>
        </td>
        <td><p>Tests if start or and end of junc is on exon.</p></td>
      </tr><tr><td>
          <p><code><a href="choose_tx.html">choose_tx()</a></code> </p>
        </td>
        <td><p>Select a subset of transcripts per junction that are more likely to be affected by a junction.</p></td>
      </tr><tr><td>
          <p><code><a href="annotate_potential_jet.html">annotate_potential_jet()</a></code> </p>
        </td>
        <td><p>Annotate if splice sites of a junction overlap with genomic positions of transposable elements.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="requantification">Requantification <a href="#requantification" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions for read support re-quantification with easyquant</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="map_requant.html">map_requant()</a></code> </p>
        </td>
        <td><p>Maps the re-quantification result from EasyQuant. on the junction-transcript centric tibble by hash id.</p></td>
      </tr><tr><td>
          <p><code><a href="read_requant.html">read_requant()</a></code> </p>
        </td>
        <td><p>Imports the re-quantification results from analysis with EasyQuant</p></td>
      </tr><tr><td>
          <p><code><a href="transform_for_requant.html">transform_for_requant()</a></code> </p>
        </td>
        <td><p>Creates a table with context sequences for re-quantification analysis with EasyQuant.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="canonical">Canonical <a href="#canonical" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions annotate and filter for cnaonical splicing</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="parse_gtf.html">parse_gtf()</a></code> </p>
        </td>
        <td><p>Parse a GFF/GTF file as <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html" class="external-link">GRangesList</a></code> of exons</p></td>
      </tr><tr><td>
          <p><code><a href="canonical_junctions.html">canonical_junctions()</a></code> </p>
        </td>
        <td><p>Build canonical junctions from transcripts</p></td>
      </tr><tr><td>
          <p><code><a href="is_canonical.html">is_canonical()</a></code> </p>
        </td>
        <td><p>test if junction is canonical junction</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="id-transformation">ID transformation <a href="#id-transformation" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to generate and transform ids</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="generate_junction_id.html">generate_junction_id()</a></code> </p>
        </td>
        <td><p>Given the chromosome, junction start, junction end and strand, a junction id is created that follows the format: <code>&lt;chr&gt;:&lt;start&gt;-&lt;end&gt;:&lt;strand&gt;</code></p></td>
      </tr><tr><td>
          <p><code><a href="breakpoint2junc.html">breakpoint2junc()</a></code> </p>
        </td>
        <td><p>Transforms breakpoint IDs (BPID) with given transcription strand into the junction ID of the format <code>&lt;chr&gt;:&lt;start&gt;-:&lt;end&gt;:&lt;strand&gt;</code></p></td>
      </tr><tr><td>
          <p><code><a href="junc2breakpoint.html">junc2breakpoint()</a></code> </p>
        </td>
        <td><p>Transforms the junction id into the breakpoint id of the format <code>&lt;chr&gt;:&lt;start&gt;-&lt;chr&gt;:&lt;end&gt;</code></p></td>
      </tr><tr><td>
          <p><code><a href="liftover_junc_id.html">liftover_junc_id()</a></code> </p>
        </td>
        <td><p>LiftOver junction IDs using the liftOver tool</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="merge-and-filter">Merge and filter <a href="#merge-and-filter" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions annotate and filter for canonical splicing</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="unique_mut_junc.html">unique_mut_junc()</a></code> </p>
        </td>
        <td><p>Filter junctions for unique combinations of mutation, transcript, and junctions.</p></td>
      </tr><tr><td>
          <p><code><a href="is_in_rnaseq.html">is_in_rnaseq()</a></code> </p>
        </td>
        <td><p>test if junction was found in corresponding RNA-seq data</p></td>
      </tr><tr><td>
          <p><code><a href="combine_mut_junc.html">combine_mut_junc()</a></code> </p>
        </td>
        <td><p>Combines data sets with junctions from several different sources</p></td>
      </tr><tr><td>
          <p><code><a href="add_identified_in_RNA.html">add_identified_in_RNA()</a></code> </p>
        </td>
        <td><p>This is a wrapper function to directly map the information if a junction predicted from WES data was found in RNA-seq by Regtools or SplAdder</p></td>
      </tr><tr><td>
          <p><code><a href="generate_combined_dataset.html">generate_combined_dataset()</a></code> </p>
        </td>
        <td><p>Combines tibbles with junctions from any number of RNA-seq tools into a combined dataset of expressed splice junctions</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="others">Others <a href="#others" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Other functions</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="junc_to_gr.html">junc_to_gr()</a></code> </p>
        </td>
        <td><p>Convert a splice junction ID into a GRanges object</p></td>
      </tr><tr><td>
          <p><code><a href="get_intronretention_alt_pos.html">get_intronretention_alt_pos()</a></code> </p>
        </td>
        <td><p>Get the alternative position of input junction from an intron retention event in the transcript sequences</p></td>
      </tr><tr><td>
          <p><code><a href="next_junctions.html">next_junctions()</a></code> </p>
        </td>
        <td><p>Get genomic coordinates of possible next donor and acceptor sides.</p></td>
      </tr><tr><td>
          <p><code><a href="sort_columns.html">sort_columns()</a></code> </p>
        </td>
        <td><p>Sorts columns of junction output file in the following order: "junction_start", "junction_end", "strand", "chromosome", "Gene", "class", "junction_id"</p></td>
      </tr><tr><td>
          <p><code><a href="unsorted_junc_df.html">unsorted_junc_df</a></code> </p>
        </td>
        <td><p>A tibble in junction format but unsorted columns</p></td>
      </tr><tr><td>
          <p><code><a href="pangolin_effect_translation.html">pangolin_effect_translation</a></code> </p>
        </td>
        <td><p>This dataset translates the increase/decrease splicing score from pangolin into a donor gain/loss and acceptor gain/loss effect annotations.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="data">Data <a href="#data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Toy example datasets</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="toy_cds.html">toy_cds</a></code> </p>
        </td>
        <td><p>An example CDS annotation GRangesList from a subset of human genome</p></td>
      </tr><tr><td>
          <p><code><a href="toy_junc_df.html">toy_junc_df</a></code> </p>
        </td>
        <td><p>An example dataset of 18 splice junctions from human (hg19) genome as a data.frame</p></td>
      </tr><tr><td>
          <p><code><a href="toy_junc_id.html">toy_junc_id</a></code> </p>
        </td>
        <td><p>An example dataset of 18 splice junctions in <code>junc_id</code> format from human (hg19) genome as character vector</p></td>
      </tr><tr><td>
          <p><code><a href="toy_junc_id_enst.html">toy_junc_id_enst</a></code> </p>
        </td>
        <td><p>An example dataset of 18 transcript IDs matching to the junctions in <code>toy_junc_id</code></p></td>
      </tr><tr><td>
          <p><code><a href="toy_transcripts.html">toy_transcripts</a></code> </p>
        </td>
        <td><p>An example transcript annotation GRangesList from a subset of human genome</p></td>
      </tr><tr><td>
          <p><code><a href="toy_transcripts_gr.html">toy_transcripts_gr</a></code> </p>
        </td>
        <td><p>An example dataset of full transcript ranges from a subset of human genome.</p></td>
      </tr><tr><td>
          <p><code><a href="unsorted_junc_df.html">unsorted_junc_df</a></code> </p>
        </td>
        <td><p>A tibble in junction format but unsorted columns</p></td>
      </tr><tr><td>
          <p><code><a href="canonical_juncs.html">canonical_juncs</a></code> </p>
        </td>
        <td><p>A tibble with canonical junctions and their source (comma, separated). example data</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jonas Ibn-Salem, Franziska Lang, Michelle Cervante, Johannes Hausmann, Patrick Sorn.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

