<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Splice2neo example analysis • splice2neo</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Splice2neo example analysis">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">splice2neo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.6.14</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/splice2neo_workflow.html">Splice2neo example analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/TRON-Bioinformatics/splice2neo/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Splice2neo example analysis</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/TRON-Bioinformatics/splice2neo/blob/v0.6.14/vignettes/splice2neo_workflow.Rmd" class="external-link"><code>vignettes/splice2neo_workflow.Rmd</code></a></small>
      <div class="hidden name"><code>splice2neo_workflow.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The R package splice2neo provides multiple functions for the analysis
of splice junctions. The main focus is on the association of splice
junctions with somatic mutations. The package integrates the output of
several tools that predict splicing effects from mutations. It annotates
these mutation effects with the resulting splicing event (alternative
3’/5’ splice sites, exons skipping, intron retentions) and the resulting
splice junctions in a standardized format. Expressed splice junctions
detected by external tools from RNA-seq data can be parsed and
integrated. Splice junctions can be filtered against canonical splice
junctions and annotated with the altered transcript sequences, CDS, and
resulting peptide sequences. Splice2neo was mainly developed to support
the detection of neoantigen candidates from tumor-specific splice
junctions. Neoantigens are tumor-specific mutated peptides recognized by
T-cells and utilized in immunotherapies against tumors <span class="citation">(Lang et al. 2022)</span>. However, most functions from
the splice2neo package are also useful for other types of splicing
analysis.</p>
<p>In this vignette, we use use some toy example data to demonstrate how
splice2neo can be used to analyze splice-junctions which are derived
from mutation predictions as tumor-specific neoantigen candidates <span class="citation">(Lang et al. 2023)</span>.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The R package splice2neo is not yet on <a href="https://CRAN.R-project.org" class="external-link">CRAN</a> or <a href="https://www.bioconductor.org/" class="external-link">Bioconductor</a>. Therefore, you
need to install splice2neo from <a href="https://github.com/TRON-Bioinformatics/splice2neo" class="external-link">github</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"remotes"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"TRON-Bioinformatics/splice2neo"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="example-workflow-to-predict-mutation-associated-alternative-splicing">Example workflow to predict mutation-associated alternative
splicing<a class="anchor" aria-label="anchor" href="#example-workflow-to-predict-mutation-associated-alternative-splicing"></a>
</h2>
<p>Integrating splice2neo functions and detection rules based on splice
effect scores and RNA-seq support facilitates the identification of
mutation-associated splice junctions which are tumor-specific and can
encode neoantigen candidates. The following workflow explains how to use
splice2neo to predict neoantigen candidates from mutation-associated
splice junctions using example data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/TRON-Bioinformatics/splice2neo" class="external-link">splice2neo</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="building-transcript-database">Building transcript database<a class="anchor" aria-label="anchor" href="#building-transcript-database"></a>
</h3>
<p>A database of transcripts is required for transcript annotation with
splice2neo. Here, we show how to build the database from a GTF file that
is a subset of GENCODE annotations, but users can build these databases
also from other resources.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># use gtf file of choice and transform into transcript database</span></span>
<span><span class="va">gtf_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example/gencode.v34lift37.annotation.subset.gtf.gz"</span>, package<span class="op">=</span><span class="st">"splice2neo"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># parse GTF file as txdb object</span></span>
<span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu">txdbmaker</span><span class="fu">::</span><span class="fu">makeTxDbFromGFF</span><span class="op">(</span><span class="va">gtf_file</span><span class="op">)</span></span>
<span><span class="co">#&gt; Import genomic features from the file as a GRanges object ... OK</span></span>
<span><span class="co">#&gt; Prepare the 'metadata' data frame ... OK</span></span>
<span><span class="co">#&gt; Make the TxDb object ...</span></span>
<span><span class="co">#&gt; Warning in .get_cds_IDX(mcols0$type, mcols0$phase): The "phase" metadata column contains non-NA values for features of type</span></span>
<span><span class="co">#&gt;   stop_codon. This information was ignored.</span></span>
<span><span class="co">#&gt; Warning in .makeTxDb_normarg_chrominfo(chrominfo): genome version information</span></span>
<span><span class="co">#&gt; is not available for this TxDb object</span></span>
<span><span class="co">#&gt; OK</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># optionally save database for later re-use </span></span>
<span><span class="co"># saveDb(txdb, file = "/path/to/transripts/txdb.sqlite")</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Build transcript database</span></span>
<span><span class="va">transcripts</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">GenomicFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcriptsBy.html" class="external-link">exonsBy</a></span><span class="op">(</span><span class="va">txdb</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tx"</span><span class="op">)</span>, use.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">transcripts_gr</span> <span class="op">&lt;-</span> <span class="fu">GenomicFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html" class="external-link">transcripts</a></span><span class="op">(</span><span class="va">txdb</span>, columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene_id"</span>, <span class="st">"tx_id"</span>, <span class="st">"tx_name"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Build cds database</span></span>
<span><span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">GenomicFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcriptsBy.html" class="external-link">cdsBy</a></span><span class="op">(</span><span class="va">txdb</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tx"</span><span class="op">)</span>, use.name <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="canonical-junctions">Canonical junctions<a class="anchor" aria-label="anchor" href="#canonical-junctions"></a>
</h3>
<p>We can retrieve the canonical splice junctions between all adjacent
exons from a list of transcripts with
<code><a href="../reference/canonical_junctions.html">canonical_junctions()</a></code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ref_junc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/canonical_junctions.html">canonical_junctions</a></span><span class="op">(</span><span class="va">transcripts</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">ref_junc</span><span class="op">)</span></span>
<span><span class="co">#&gt;  chr [1:1316] "chr2:28680064-28739675:+" "chr2:28739736-28741333:+" ...</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="splice-junction-id">Splice junction id<a class="anchor" aria-label="anchor" href="#splice-junction-id"></a>
</h3>
<p>In splice2neo, splice junctions are represented by a
<code>junc_id</code>, which is defined by the genomic coordinates of the
exon positions (1-based) that are combined through the junction and the
transcription strand of the transcript:
<code>chr:pos1-pos2:strand</code>.</p>
<p>This format allows direct conversion to <code>GRangs</code> objects
from the <a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a>
package.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="va">ref_junc</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gr</span></span>
<span><span class="co">#&gt; GRanges object with 1316 ranges and 0 metadata columns:</span></span>
<span><span class="co">#&gt;          seqnames            ranges strand</span></span>
<span><span class="co">#&gt;             &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">#&gt;      [1]     chr2 28680064-28739675      +</span></span>
<span><span class="co">#&gt;      [2]     chr2 28739736-28741333      +</span></span>
<span><span class="co">#&gt;      [3]     chr2 28741399-28742572      +</span></span>
<span><span class="co">#&gt;      [4]     chr2 28742630-28748134      +</span></span>
<span><span class="co">#&gt;      [5]     chr2 28748174-28748772      +</span></span>
<span><span class="co">#&gt;      ...      ...               ...    ...</span></span>
<span><span class="co">#&gt;   [1312]     chrX 76949322-76949323      -</span></span>
<span><span class="co">#&gt;   [1313]     chrX 76944394-76944395      -</span></span>
<span><span class="co">#&gt;   [1314]     chrX 76952175-76952176      -</span></span>
<span><span class="co">#&gt;   [1315]     chrX 77041690-77041691      -</span></span>
<span><span class="co">#&gt;   [1316]     chrX 76953090-76953091      -</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 13 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>A given GRanges object with ranges of splice junctions can be
converted into a vector of `junc_id``s:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">junc_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">junc_id</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "chr2:28680064-28739675:+" "chr2:28739736-28741333:+"</span></span>
<span><span class="co">#&gt; [3] "chr2:28741399-28742572:+" "chr2:28742630-28748134:+"</span></span>
<span><span class="co">#&gt; [5] "chr2:28748174-28748772:+" "chr2:28748812-28752184:+"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="annotation-of-rna-seq-derived-splice-junctions">Annotation of RNA-seq derived splice junctions<a class="anchor" aria-label="anchor" href="#annotation-of-rna-seq-derived-splice-junctions"></a>
</h3>
<p>Splice2neo transforms splice events identified in RNA-seq into a
standardized junction format. We show here how to parse the tools
LeafCutter and SplAdder.</p>
<div class="section level4">
<h4 id="leafcutter">LeafCutter<a class="anchor" aria-label="anchor" href="#leafcutter"></a>
</h4>
<p>The output of the tool <a href="https://github.com/davidaknowles/leafcutter" class="external-link">LeafCutter</a> <span class="citation">(Li et al. 2018)</span> can be transformed using
<code><a href="../reference/leafcutter_transform.html">leafcutter_transform()</a></code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leafcutter_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example/leafcutter"</span>, package<span class="op">=</span><span class="st">"splice2neo"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">leafcutter</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/leafcutter_transform.html">leafcutter_transform</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">leafcutter_path</span><span class="op">)</span></span>
<span><span class="co">#&gt; Importing Leafcutter files...</span></span>
<span><span class="co">#&gt; Importing Leafcutter files completed</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">leafcutter</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 7</span></span></span>
<span><span class="co">#&gt;   junction_start junction_end strand chromosome Gene  class junc_id             </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 39934433       39937097     -      chrX       <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    chrX:39934433-39937…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 76944420       76953071     -      chrX       <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    chrX:76944420-76953…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 18026082       18029488     -      chr11      <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    chr11:18026082-1802…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 20921104       20922809     +      chr22      <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    chr22:20921104-2092…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 153333024      153456093    -      chr4       <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    chr4:153333024-1534…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 32084907       32085100     -      chr6       <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    chr6:32084907-32085…</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="spladder">SplAdder<a class="anchor" aria-label="anchor" href="#spladder"></a>
</h4>
<p>The output of the tool <a href="https://github.com/ratschlab/spladder" class="external-link">SplAdder</a> <span class="citation">(Kahles et al. 2016)</span> can be transformed using
<code><a href="../reference/spladder_transform.html">spladder_transform()</a></code>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spladder_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example/spladder"</span>, package<span class="op">=</span><span class="st">"splice2neo"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">spladder</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/spladder_transform.html">spladder_transform</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">spladder_path</span><span class="op">)</span></span>
<span><span class="co">#&gt; Importing Spladder files ...</span></span>
<span><span class="co">#&gt; Importing Spladder files completed</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">spladder</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 8</span></span></span>
<span><span class="co">#&gt;   junction_start junction_end strand chromosome Gene               class junc_id</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 20921104       20922808     +      chr22      ENSG00000099917.1… A3SS  chr22:…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 20921104       20922809     +      chr22      ENSG00000099917.1… A3SS  chr22:…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 41108521       41109155     -      chr17      ENSG00000266967.7… A3SS  chr17:…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 41108577       41109155     -      chr17      ENSG00000266967.7… A3SS  chr17:…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 148881736      148884821    +      chr3       ENSG00000163755.9… A3SS  chr3:1…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 148881736      148884852    +      chr3       ENSG00000163755.9… A3SS  chr3:1…</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: number_supporting_reads &lt;dbl&gt;</span></span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="combine-rna-seq-derived-splice-junctions">Combine RNA-seq derived splice junctions<a class="anchor" aria-label="anchor" href="#combine-rna-seq-derived-splice-junctions"></a>
</h4>
<p>Now we want to get a single <code>data.frame</code> of all splice
junctions identified by RNA-seq tools and information by which tool they
were identified.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_junctions</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/generate_combined_dataset.html">generate_combined_dataset</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"spladder_juncs"</span> <span class="op">=</span> <span class="va">spladder</span>, <span class="st">"leafcutter_juncs"</span> <span class="op">=</span> <span class="va">leafcutter</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">rna_junctions</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 10</span></span></span>
<span><span class="co">#&gt;   junc_id   junction_start junction_end strand chromosome identified_by_leafcu…¹</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>                 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr1:266… 26627515       26628185     -      chr1       TRUE                  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr1:849… 84964231       84971694     -      chr1       TRUE                  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> chr10:82… 82267130       82269057     +      chr10      FALSE                 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> chr10:82… 82267130       82271900     +      chr10      FALSE                 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> chr10:82… 82269227       82271900     +      chr10      FALSE                 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> chr11:18… 18026082       18029488     -      chr11      TRUE                  </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ abbreviated name: ¹​identified_by_leafcutter_juncs</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 4 more variables: identified_by_spladder_juncs &lt;lgl&gt;, Gene &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   class &lt;chr&gt;, spladder_juncs_number_supporting_reads &lt;dbl&gt;</span></span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="annotation-of-mutation-effect-on-splicing">Annotation of mutation effect on splicing<a class="anchor" aria-label="anchor" href="#annotation-of-mutation-effect-on-splicing"></a>
</h3>
<p>Deep learning tools can predict the effect of a mutation on the
disruption or creation of canonical splicing sequence motifs. This is
typically provided as gain or loss of splicing acceptor or donor sites.
Splice2neo can transform and annotate such mutation effects into a
unified splice junction format. Currently, the transformation of
predictions by <a href="https://github.com/Illumina/SpliceAI" class="external-link">SpliceAI</a> <span class="citation">(Jaganathan et al. 2019)</span>, <a href="https://github.com/gagneurlab/MMSplice_MTSplice" class="external-link">MMSplice</a><span class="citation">(Cheng et al. 2019)</span> and <a href="https://github.com/tkzeng/Pangolin" class="external-link">Pangolin</a><span class="citation">(Zeng and Li 2022)</span> are supported.<br>
Here, we show how to annotate results from SpliceAI and MMSplice. The
annotation of Pangolin results works analogously with SpliceAI.</p>
<div class="section level4">
<h4 id="spliceai">SpliceAI<a class="anchor" aria-label="anchor" href="#spliceai"></a>
</h4>
<p>Effect predictions by SpliceAI can be read with
<code><a href="../reference/parse_spliceai.html">parse_spliceai()</a></code></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spliceai_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example/spliceAI.vcf"</span>, package<span class="op">=</span><span class="st">"splice2neo"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># parse SpliceAI results </span></span>
<span><span class="va">spliceai</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/parse_spliceai.html">parse_spliceai</a></span><span class="op">(</span><span class="va">spliceai_file</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">spliceai</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 18</span></span></span>
<span><span class="co">#&gt;   CHROM POS      ID    REF   ALT   QUAL  FILTER   Key ALLELE SYMBOL  DS_AG DS_AL</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr10 8006689  <span style="color: #BB0000;">NA</span>    C     T     <span style="color: #BB0000;">NA</span>    PASS       1 T      TAF3     0     0   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr10 91528535 <span style="color: #BB0000;">NA</span>    G     A     <span style="color: #BB0000;">NA</span>    PASS       2 A      KIF20B   0.64  0.22</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> chr11 18028133 <span style="color: #BB0000;">NA</span>    C     T     <span style="color: #BB0000;">NA</span>    PASS       3 T      SERGEF   0     0   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> chr11 62985208 <span style="color: #BB0000;">NA</span>    C     T     <span style="color: #BB0000;">NA</span>    PASS       4 T      SLC22A…  0.06  0.55</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> chr16 27709648 <span style="color: #BB0000;">NA</span>    G     A     <span style="color: #BB0000;">NA</span>    PASS       5 A      KIAA05…  0.97  1   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> chr16 30020302 <span style="color: #BB0000;">NA</span>    G     A     <span style="color: #BB0000;">NA</span>    PASS       6 A      DOC2A    0     0   </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 6 more variables: DS_DG &lt;dbl&gt;, DS_DL &lt;dbl&gt;, DP_AG &lt;int&gt;, DP_AL &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   DP_DG &lt;int&gt;, DP_DL &lt;int&gt;</span></span></span></code></pre></div>
<p>Next, SpliceAi effects can be transformed into a standardized effect
dataframe with <code><a href="../reference/format_spliceai.html">format_spliceai()</a></code>. Here, we can optionally
provide a <code>data.frame</code> (<code>gene_table</code>) which
relates gene symbols with GENCODE gene ids. If <code>gene_table</code>
is provided, the formatted <code>data.frame</code> contains the GENCODE
gene ids. While this is advised to better select relevant transcripts,
it is not required.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_table</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example/gene_table.tsv"</span>, package <span class="op">=</span></span>
<span>                                <span class="st">"splice2neo"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">17</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">3</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">chr</span> (3): strand, gene_id, gene_name</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gene_table</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span></span>
<span><span class="co">#&gt;   strand gene_id              gene_name</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> -      ENSG00000134313.16_8 KIDINS220</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> +      ENSG00000163803.13_8 PLB1     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> -      ENSG00000163510.14_5 CWC22    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> -      ENSG00000136237.19_8 RAPGEF5  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> +      ENSG00000147316.13_4 MCPH1    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> +      ENSG00000138182.14_3 KIF20B</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># format SpliceAI results</span></span>
<span><span class="va">splicai_formatted</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/format_spliceai.html">format_spliceai</a></span><span class="op">(</span><span class="va">spliceai</span>, gene_table <span class="op">=</span> <span class="va">gene_table</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">splicai_formatted</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 7</span></span></span>
<span><span class="co">#&gt;   mut_id             effect score chr   pos_rel      pos gene_id             </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr10_91528535_G_A AG      0.64 chr10       2 91<span style="text-decoration: underline;">528</span>537 ENSG00000138182.14_3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr10_91528535_G_A AL      0.22 chr10     -<span style="color: #BB0000;">40</span> 91<span style="text-decoration: underline;">528</span>495 ENSG00000138182.14_3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> chr11_18028133_C_T DG      0.09 chr11      42 18<span style="text-decoration: underline;">028</span>175 ENSG00000129158.11_6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> chr11_18028133_C_T DL      0.15 chr11       5 18<span style="text-decoration: underline;">028</span>138 ENSG00000129158.11_6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> chr11_62985208_C_T AG      0.06 chr11      -<span style="color: #BB0000;">2</span> 62<span style="text-decoration: underline;">985</span>206 ENSG00000196600.12_6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> chr11_62985208_C_T AL      0.55 chr11      -<span style="color: #BB0000;">1</span> 62<span style="text-decoration: underline;">985</span>207 ENSG00000196600.12_6</span></span></code></pre></div>
<p>Predicted SpliceAI effects can be annotated with all potential
resulting junctions and transcripts with
<code><a href="../reference/annotate_mut_effect.html">annotate_mut_effect()</a></code>. If the input was formatted using
<code>gene_table</code>, the user may want to choose
<code>gene_mapping = TRUE</code> which removes unlikely events from the
results. Otherwise, the default <code>gene_mapping = FALSE</code> should
be used. The resulting dataframe contains all potential alternative
transcripts per row, meaning that there can be multiple annotated
transcripts per junction and also multiple junctions per mutation
effect. At this point, no filtering based on effect size has been
applied meaning that the dataframe contains unlikely events.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># annotate SpliceAI junctions with all potential junctions and transcripts</span></span>
<span><span class="va">spliceai_annotated</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/annotate_mut_effect.html">annotate_mut_effect</a></span><span class="op">(</span></span>
<span>    effect_df <span class="op">=</span> <span class="va">splicai_formatted</span>,</span>
<span>    transcripts <span class="op">=</span> <span class="va">transcripts</span>,</span>
<span>    transcripts_gr <span class="op">=</span> <span class="va">transcripts_gr</span>,</span>
<span>    gene_mapping <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; INFO: calculate coordinates of upstream and downstream exons...</span></span>
<span><span class="co">#&gt; INFO: get overlapping transcripts...</span></span>
<span><span class="co">#&gt; INFO: Build df ...</span></span>
<span><span class="co">#&gt; INFO: build pairs...</span></span>
<span><span class="co">#&gt; INFO: find overlapping exons...</span></span>
<span><span class="co">#&gt; INFO: calculate junction coordinates from predicted effect...</span></span>
<span><span class="co">#&gt; INFO: Evaluation of rules done.</span></span>
<span></span>
<span></span>
<span><span class="va">spliceai_annotated</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">mut_id</span>, <span class="va">junc_id</span>, <span class="va">gene_id</span>, <span class="va">tx_id</span>, <span class="va">effect</span>, <span class="va">score</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 6</span></span></span>
<span><span class="co">#&gt;   mut_id             junc_id                   gene_id        tx_id effect score</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr10_91528535_G_A chr10:91528148-91528537:+ ENSG000001381… ENST… AG      0.64</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr10_91528535_G_A chr10:91528148-91528537:+ ENSG000001381… ENST… AG      0.64</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> chr10_91528535_G_A chr10:91528148-91528537:+ ENSG000001381… ENST… AG      0.64</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> chr10_91528535_G_A chr10:91528494-91528495:+ ENSG000001381… ENST… AL      0.22</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> chr10_91528535_G_A chr10:91528148-91532446:+ ENSG000001381… ENST… AL      0.22</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> chr10_91528535_G_A chr10:91528494-91528495:+ ENSG000001381… ENST… AL      0.22</span></span></code></pre></div>
<p>It can happen that multiple effects lead to the same altered
transcripts (defined by <code>mut_id</code>, <code>junc_id</code>,
<code>tx_id</code> ). The function <code><a href="../reference/unique_mut_junc.html">unique_mut_junc()</a></code> can be
used to filter for unique altered transcripts by keeping the effect with
the highest effect score.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># unique altered transcripts</span></span>
<span><span class="va">spliceai_annotated_unique</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/unique_mut_junc.html">unique_mut_junc</a></span><span class="op">(</span><span class="va">spliceai_annotated</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">spliceai_annotated_unique</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">spliceai_annotated</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="mmsplice">MMSplice<a class="anchor" aria-label="anchor" href="#mmsplice"></a>
</h4>
<p>We can read the effect predictions by MMSplice with
<code><a href="../reference/parse_mmsplice.html">parse_mmsplice()</a></code></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mmsplice_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example/MMsplice.csv"</span>, package<span class="op">=</span><span class="st">"splice2neo"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># parse MMSplice results </span></span>
<span><span class="va">mmsplice</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/parse_mmsplice.html">parse_mmsplice</a></span><span class="op">(</span><span class="va">mmsplice_file</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mmsplice</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 20</span></span></span>
<span><span class="co">#&gt;   ID               exons exon_id gene_id gene_name transcript_id delta_logit_psi</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr2:28764482:C… chr2… ENSE00… ENSG00… PLB1      ENST00000327…         0.004<span style="text-decoration: underline;">44</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr2:28764482:C… chr2… ENSE00… ENSG00… PLB1      ENST00000422…         0.004<span style="text-decoration: underline;">44</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> chr2:28764482:C… chr2… ENSE00… ENSG00… PLB1      ENST00000404…         0.004<span style="text-decoration: underline;">44</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> chr2:97312723:C… chr2… ENSE00… ENSG00… FER1L5    ENST00000624…         0.046<span style="text-decoration: underline;">4</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> chr2:97312723:C… chr2… ENSE00… ENSG00… FER1L5    ENST00000436…         0.046<span style="text-decoration: underline;">4</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> chr2:97312723:C… chr2… ENSE00… ENSG00… FER1L5    ENST00000641…         0.046<span style="text-decoration: underline;">4</span> </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 13 more variables: ref_acceptorIntron &lt;dbl&gt;, ref_acceptor &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ref_exon &lt;dbl&gt;, ref_donor &lt;dbl&gt;, ref_donorIntron &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   alt_acceptorIntron &lt;dbl&gt;, alt_acceptor &lt;dbl&gt;, alt_exon &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   alt_donor &lt;dbl&gt;, alt_donorIntron &lt;dbl&gt;, pathogenicity &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   efficiency &lt;dbl&gt;, mut_id &lt;chr&gt;</span></span></span></code></pre></div>
<p>The MMSplice data can be annotated with all potential splice
junctions using <code><a href="../reference/annotate_mmsplice.html">annotate_mmsplice()</a></code></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mmsplice_annotated</span> <span class="op">&lt;-</span> <span class="va">mmsplice</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/annotate_mmsplice.html">annotate_mmsplice</a></span><span class="op">(</span>transcripts <span class="op">=</span> <span class="va">transcripts</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mmsplice_annotated</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 22</span></span></span>
<span><span class="co">#&gt;   ID    exons exon_id gene_id gene_name tx_id delta_logit_psi ref_acceptorIntron</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr2… chr2… ENSE00… ENSG00… PLB1      ENST…         0.004<span style="text-decoration: underline;">44</span>              -<span style="color: #BB0000;">4.31</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr2… chr2… ENSE00… ENSG00… PLB1      ENST…         0.004<span style="text-decoration: underline;">44</span>              -<span style="color: #BB0000;">4.31</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> chr2… chr2… ENSE00… ENSG00… PLB1      ENST…         0.004<span style="text-decoration: underline;">44</span>              -<span style="color: #BB0000;">4.31</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> chr2… chr2… ENSE00… ENSG00… PLB1      ENST…         0.004<span style="text-decoration: underline;">44</span>              -<span style="color: #BB0000;">4.31</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> chr2… chr2… ENSE00… ENSG00… PLB1      ENST…         0.004<span style="text-decoration: underline;">44</span>              -<span style="color: #BB0000;">4.31</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> chr2… chr2… ENSE00… ENSG00… PLB1      ENST…         0.004<span style="text-decoration: underline;">44</span>              -<span style="color: #BB0000;">4.31</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 14 more variables: ref_acceptor &lt;dbl&gt;, ref_exon &lt;dbl&gt;, ref_donor &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ref_donorIntron &lt;dbl&gt;, alt_acceptorIntron &lt;dbl&gt;, alt_acceptor &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   alt_exon &lt;dbl&gt;, alt_donor &lt;dbl&gt;, alt_donorIntron &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   pathogenicity &lt;dbl&gt;, efficiency &lt;dbl&gt;, mut_id &lt;chr&gt;, event_type &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   junc_id &lt;chr&gt;</span></span></span></code></pre></div>
<p>Junctions predicted from MMSplice results can contain duplicated
altered transcripts for exon inclusion events. We use the function
<code><a href="../reference/unique_junc_mmsplice.html">unique_junc_mmsplice()</a></code> to filter for unique altered
transcripts by keeping the effect with the highest effect score.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># unique altered transcripts</span></span>
<span><span class="va">mmsplice_annotated_unique</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/unique_junc_mmsplice.html">unique_junc_mmsplice</a></span><span class="op">(</span><span class="va">mmsplice_annotated</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">mmsplice_annotated_unique</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">mmsplice_annotated</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="combine-mutation-retrieved-splice-junctions">Combine mutation-retrieved splice junctions<a class="anchor" aria-label="anchor" href="#combine-mutation-retrieved-splice-junctions"></a>
</h4>
<p>Next, we combine splice junctions from SpliceAI and MMSplice into one
dataframe using <code><a href="../reference/combine_mut_junc.html">combine_mut_junc()</a></code>. This dataframe will
contain unique altered transcripts (defined by <code>mut_id</code>,
<code>junc_id</code>, <code>tx_id</code>) in rows. Tool specific
information are defined in columns starting with the respective tool
name.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mut_junctions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combine_mut_junc.html">combine_mut_junc</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"spliceai"</span> <span class="op">=</span> <span class="va">spliceai_annotated_unique</span>, </span>
<span>    <span class="st">"mmsplice"</span> <span class="op">=</span> <span class="va">mmsplice_annotated_unique</span></span>
<span>    <span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">mut_junctions</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">mut_id</span>, <span class="va">junc_id</span>, <span class="va">tx_id</span>, <span class="va">spliceai_event_type</span>, <span class="va">spliceai_score</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 5</span></span></span>
<span><span class="co">#&gt;   mut_id             junc_id            tx_id spliceai_event_type spliceai_score</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr10_8006689_C_T  chr10:7866523-801… ENST… <span style="color: #BB0000;">NA</span>                           <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr10_91528535_G_A chr10:91528148-91… ENST… alternative 3prim             0.64</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> chr10_91528535_G_A chr10:91528148-91… ENST… exon skipping                 0.22</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> chr10_91528535_G_A chr10:91528494-91… ENST… intron retention              0.22</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> chr10_91528535_G_A chr10:91528148-91… ENST… alternative 3prim             0.64</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> chr10_91528535_G_A chr10:91528148-91… ENST… exon skipping                 0.22</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="integrative-analysis-of-mutation-retrieved-splice-junctions">Integrative analysis of mutation-retrieved splice junctions<a class="anchor" aria-label="anchor" href="#integrative-analysis-of-mutation-retrieved-splice-junctions"></a>
</h3>
<div class="section level4">
<h4 id="annotate-canonical-junctions">Annotate canonical junctions<a class="anchor" aria-label="anchor" href="#annotate-canonical-junctions"></a>
</h4>
<p>To exclude junctions expressed in healthy tissues, we here want to
filter predicted junctions for non-canonical splice junctions. We can
use for instance exon-exon junctions from GENCODE and junctions found in
healthy tisssue samples. Splice2neo also provides the function
<code><a href="../reference/bed_to_junc.html">bed_to_junc()</a></code> which allows to transform a bed file into a
list of splice junctions.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this is just an example file with a few canonical junctions --&gt; incomplete</span></span>
<span><span class="va">canonical_juncs_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example/test_canonical.bed"</span>, package<span class="op">=</span><span class="st">"splice2neo"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">canonical_junctios</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bed_to_junc.html">bed_to_junc</a></span><span class="op">(</span>bed_file <span class="op">=</span> <span class="va">canonical_juncs_file</span>, type <span class="op">=</span> <span class="st">"exon-exon"</span><span class="op">)</span></span></code></pre></div>
<p>Annotate mutation-retrieved splice junction whether they are
canonical splice junctions.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># annotated canonical junctions</span></span>
<span><span class="va">mut_junctions</span> <span class="op">&lt;-</span> <span class="va">mut_junctions</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>is_canonical <span class="op">=</span> <span class="fu"><a href="../reference/is_canonical.html">is_canonical</a></span><span class="op">(</span><span class="va">junc_id</span>, ref_junc <span class="op">=</span> <span class="va">canonical_junctios</span>, exons_gr <span class="op">=</span> <span class="va">transcripts</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mut_junctions</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">is_canonical</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 2</span></span></span>
<span><span class="co">#&gt;   is_canonical     n</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> FALSE          202</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> TRUE           124</span></span>
<span></span>
<span><span class="co"># remove canonical junctions </span></span>
<span><span class="va">noncan_mut_junctions</span> <span class="op">&lt;-</span> <span class="va">mut_junctions</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">is_canonical</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="annotate-support-in-rna-seq-derived-splice-junctions">Annotate support in RNA-seq derived splice junctions<a class="anchor" aria-label="anchor" href="#annotate-support-in-rna-seq-derived-splice-junctions"></a>
</h4>
<p>Annotate whether mutation-retrieved splice junction were also
detected by a RNA-seq tool.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">noncan_mut_junctions</span> <span class="op">&lt;-</span> <span class="va">noncan_mut_junctions</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>is_in_rnaseq <span class="op">=</span> <span class="fu"><a href="../reference/is_in_rnaseq.html">is_in_rnaseq</a></span><span class="op">(</span><span class="va">junc_id</span>, rna_juncs <span class="op">=</span> <span class="va">rna_junctions</span><span class="op">$</span><span class="va">junc_id</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="annotate-transcript-context-sequence">Annotate transcript context sequence<a class="anchor" aria-label="anchor" href="#annotate-transcript-context-sequence"></a>
</h4>
<p>Splice2neo can annotate the resulting transcript sequence for each
altered transcript with <code><a href="../reference/add_context_seq.html">add_context_seq()</a></code>. Here, we set the
<code>size</code> parameter to 400 to extract the sequence of +/-200
nucleotides around the splice junctions as context sequence.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">noncan_mut_junctions</span> <span class="op">&lt;-</span> <span class="va">noncan_mut_junctions</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/add_context_seq.html">add_context_seq</a></span><span class="op">(</span></span>
<span>    size <span class="op">=</span> <span class="fl">400</span>,</span>
<span>    bsg <span class="op">=</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg19</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html" class="external-link">BSgenome.Hsapiens.UCSC.hg19</a></span>,</span>
<span>    transcripts <span class="op">=</span> <span class="va">transcripts</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">noncan_mut_junctions</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">mut_id</span>, <span class="va">junc_id</span>, <span class="va">tx_id</span>, <span class="va">cts_id</span>, <span class="va">cts_seq</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 5</span></span></span>
<span><span class="co">#&gt;   mut_id             junc_id                   tx_id              cts_id cts_seq</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr10_8006689_C_T  chr10:7866523-8019204:+   ENST00000344293.6… dd63d… CCAGCT…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr10_91528535_G_A chr10:91528148-91528537:+ ENST00000260753.8… 83411… AGCCAA…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> chr10_91528535_G_A chr10:91528148-91532446:+ ENST00000260753.8… f7ae9… AGCCAA…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> chr10_91528535_G_A chr10:91528494-91528495:+ ENST00000260753.8… 30ff5… AGCCAA…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> chr10_91528535_G_A chr10:91528148-91528537:+ ENST00000371728.7… 83411… AGCCAA…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> chr10_91528535_G_A chr10:91528148-91532446:+ ENST00000371728.7… f7ae9… AGCCAA…</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="annotate-peptide-context-sequence">Annotate peptide context sequence<a class="anchor" aria-label="anchor" href="#annotate-peptide-context-sequence"></a>
</h4>
<p>Splice2neo can annotate the resulting peptide sequence for each
altered transcript with <code><a href="../reference/add_peptide.html">add_peptide()</a></code>. This is relevant to
analyze which altered transcript could qualify as neoantigens. We can
define by the parameter <code>flanking_size</code> the number of
wild-type amino acids flanking the junction or novel sequence caused by
the splice junction to the left and to the right. If the junction leads
to a reading frame shift, sequences are flanked by wild-type amino acids
to the left and are annotated until the first stop codon.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">noncan_mut_junctions</span> <span class="op">&lt;-</span> <span class="va">noncan_mut_junctions</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/add_peptide.html">add_peptide</a></span><span class="op">(</span>flanking_size <span class="op">=</span> <span class="fl">13</span>,</span>
<span>              bsg <span class="op">=</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg19</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html" class="external-link">BSgenome.Hsapiens.UCSC.hg19</a></span>,</span>
<span>              cds <span class="op">=</span> <span class="va">cds</span><span class="op">)</span></span>
<span></span>
<span><span class="va">noncan_mut_junctions</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">mut_id</span>,</span>
<span>                <span class="va">junc_id</span>,</span>
<span>                <span class="va">tx_id</span>,</span>
<span>                <span class="va">junc_in_orf</span>,</span>
<span>                <span class="va">peptide_context</span>,</span>
<span>                <span class="va">frame_shift</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 6</span></span></span>
<span><span class="co">#&gt;   mut_id             junc_id       tx_id junc_in_orf peptide_context frame_shift</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr10_8006689_C_T  chr10:786652… ENST… TRUE        IPDYLPPIVSSQED… TRUE       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr10_91528535_G_A chr10:915281… ENST… TRUE        SDDRNSSVKKEQKS… FALSE      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> chr10_91528535_G_A chr10:915281… ENST… TRUE        SDDRNSSVKKEQKQ… TRUE       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> chr10_91528535_G_A chr10:915284… ENST… TRUE        SDDRNSSVKKEQKV… TRUE       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> chr10_91528535_G_A chr10:915281… ENST… TRUE        SDDRNSSVKKEQKS… FALSE      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> chr10_91528535_G_A chr10:915281… ENST… TRUE        SDDRNSSVKKEQKQ… TRUE</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="annotate-if-predicted-retained-introns-are-covered-by-an-exon-of-another-transcript">Annotate if predicted retained introns are covered by an exon of
another transcript<a class="anchor" aria-label="anchor" href="#annotate-if-predicted-retained-introns-are-covered-by-an-exon-of-another-transcript"></a>
</h4>
<p>Relevant intron retention events are particular challenging to
predict. Splice2neo provides the function <code><a href="../reference/exon_in_intron.html">exon_in_intron()</a></code>
to annotate whether the exon of another transcript is in the genomic
region of a predicted retained introns. We do not want to consider them
in the downstream analysis as it is difficult to differentiate if the
read support from RNA-seq relates to the retained intron or to the exon
of the other transcript.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># annotate exons in introns of other transcript</span></span>
<span><span class="va">noncan_mut_junctions</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">noncan_mut_junctions</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/exon_in_intron.html">exon_in_intron</a></span><span class="op">(</span>transcripts <span class="op">=</span> <span class="va">transcripts</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="re-quantification-of-altered-transcripts-with-easyquant">Re-quantification of altered transcripts with EasyQuant<a class="anchor" aria-label="anchor" href="#re-quantification-of-altered-transcripts-with-easyquant"></a>
</h4>
<p>Altered transcript from predicted splice junctions can be
re-quantified in RNA-seq in a targeted manner with EasyQuant (available
at <a href="https://github.com/TRON-Bioinformatics/easyquant" class="external-link">https://github.com/TRON-Bioinformatics/easyquant</a>).
EasyQuant maps all RNA-seq reads to to a custom reference build from the
context sequence around splice junctions. We can generate the input for
EasyQuant using the splice2neo function
<code><a href="../reference/transform_for_requant.html">transform_for_requant()</a></code>.<br>
EasyQuant should be run in interval mode for the re-quantification of
splice junction and it is advised to set the <code>BP_distannce</code>
parameter to 10 (see <a href="https://github.com/TRON-Bioinformatics/easyquant" class="external-link">EasyQuant
README</a>).</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># transform for EasyQuant</span></span>
<span><span class="va">easyquant_input</span> <span class="op">&lt;-</span> <span class="va">noncan_mut_junctions</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/transform_for_requant.html">transform_for_requant</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">easyquant_input</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span></span>
<span><span class="co">#&gt;   name                             sequence                             position</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> dd63de34c46767f447b72d5ab223df47 CCAGCTGATGGGGGTTAGTCTACATGAACTAGAAG… 200     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 834110bae711a18586e9745ce37e6d33 AGCCAAACAAAATGGCAGTGAAACACCCTGGTTGT… 200     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> f7ae934315938fe4c3b41bc44b74304c AGCCAAACAAAATGGCAGTGAAACACCCTGGTTGT… 200     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 30ff58139be454c8ca52c4a28bf8a000 AGCCAAACAAAATGGCAGTGAAACACCCTGGTTGT… 0,200,5…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 7d57cc792100728dd7c9d6cb03f0bae2 CCCAGCAACTGAATGACTTCTGTAAACCCAGGAGT… 200     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 29bc2a21dd2b55a13ddc993fc8be5086 TCGCATGGAGCGCGAGCCCAGCGCCTCGGAGGCCG… 200</span></span></code></pre></div>
<p>After running EasyQuant (which is not shown here), we can import the
EasyQuant results with the splice2neo function
<code><a href="../reference/map_requant.html">map_requant()</a></code>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># path to easyquant folder </span></span>
<span><span class="va">easyquant_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example/easyquant"</span>, package<span class="op">=</span><span class="st">"splice2neo"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># merge easyquant results </span></span>
<span><span class="va">requant_noncan_mut_junctions</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/map_requant.html">map_requant</a></span><span class="op">(</span>path_to_easyquant_folder <span class="op">=</span> <span class="va">easyquant_path</span>,</span>
<span>              junc_tib <span class="op">=</span> <span class="va">noncan_mut_junctions</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">261</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">8</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">chr</span> (2): name, interval</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">dbl</span> (6): overlap_interval_end_reads, span_interval_end_pairs, within_interva...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span></code></pre></div>
<p>Here, we are interested in different result columns depending on the
type of splice event. In case of alternative splice sites or exon
skipping events, want use the number of junction reads that map on the
splice junction (<code>junc_interval_start</code>) and/or the number of
spanning reads that bridge the splice junction
(<code>span_interval_start</code>).</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># re-quantification results</span></span>
<span><span class="va">requant_noncan_mut_junctions</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">spliceai_event_type</span> <span class="op">!=</span> <span class="st">"intron retention"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">mut_id</span>, <span class="va">junc_id</span>, <span class="va">tx_id</span>, <span class="va">junc_interval_start</span>, <span class="va">span_interval_start</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 5</span></span></span>
<span><span class="co">#&gt;   mut_id             junc_id       tx_id junc_interval_start span_interval_start</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr10_91528535_G_A chr10:915281… ENST…                   3                  10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr10_91528535_G_A chr10:915281… ENST…                   0                   9</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> chr10_91528535_G_A chr10:915281… ENST…                   3                  10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> chr10_91528535_G_A chr10:915281… ENST…                   0                   9</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> chr10_91528535_G_A chr10:915281… ENST…                   3                  10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> chr10_91528535_G_A chr10:915281… ENST…                   0                   9</span></span></code></pre></div>
<p>In case of intron retentions, we may want to consider the following
columns:</p>
<ul>
<li>
<code>within_interval</code>: Number of reads that map to the intron
of interest<br>
</li>
<li>
<code>coverage_perc</code>: Relative read coverage of the intron of
interest<br>
</li>
<li>
<code>coverage_median</code>: Median read coverage of the intron of
interest<br>
</li>
<li>
<code>coverage_mean</code>: Mean read coverage of the intron of
interest. This value can be misleading by skewed read distribution and
the user may rather want to use the median coverage</li>
</ul>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># re-quantification results</span></span>
<span><span class="va">requant_noncan_mut_junctions</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">spliceai_event_type</span> <span class="op">==</span> <span class="st">"intron retention"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>    <span class="va">mut_id</span>,</span>
<span>    <span class="va">junc_id</span>,</span>
<span>    <span class="va">tx_id</span>,</span>
<span>    <span class="va">within_interval</span>,</span>
<span>    <span class="va">coverage_perc</span>,</span>
<span>    <span class="va">coverage_median</span>,</span>
<span>    <span class="va">coverage_mean</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 7</span></span></span>
<span><span class="co">#&gt;   mut_id             junc_id tx_id within_interval coverage_perc coverage_median</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr10_91528535_G_A chr10:… ENST…               1         0.543               1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr10_91528535_G_A chr10:… ENST…               1         0.543               1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> chr10_91528535_G_A chr10:… ENST…               1         0.543               1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> chr11_18028133_C_T chr11:… ENST…              85         0.937               4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> chr11_18028133_C_T chr11:… ENST…              85         0.937               4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> chr11_18028133_C_T chr11:… ENST…              85         0.937               4</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: coverage_mean &lt;dbl&gt;</span></span></span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># alternative to only import EasyQuant results without merging </span></span>
<span></span>
<span><span class="va">requantification_results</span>  <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/read_requant.html">read_requant</a></span><span class="op">(</span>path_folder <span class="op">=</span> <span class="va">easyquant_path</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">261</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">8</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">chr</span> (2): name, interval</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">dbl</span> (6): overlap_interval_end_reads, span_interval_end_pairs, within_interva...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">requantification_results</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 20</span></span></span>
<span><span class="co">#&gt;   name                 junc_interval_start junc_interval_end span_interval_start</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 5846baef085e1cb83c8…                   0                 1                   8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> e0ab8366d4f7014c521…                   0                 1                   9</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> fcc744d7184c3bdae26…                   1                 0                   7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 13490cdd721e01c1334…                   0                 1                   3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 3f5a72af0853d8b59e6…                   2                 0                   6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 48ba45ad78650a6d3b6…                   2                 0                   3</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 16 more variables: span_interval_end &lt;dbl&gt;, within_interval &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   within_interval_left &lt;dbl&gt;, within_interval_right &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   coverage_perc &lt;dbl&gt;, coverage_perc_left &lt;dbl&gt;, coverage_perc_right &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   coverage_mean &lt;dbl&gt;, coverage_mean_left &lt;dbl&gt;, coverage_mean_right &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   coverage_median &lt;dbl&gt;, coverage_median_left &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   coverage_median_right &lt;dbl&gt;, interval &lt;chr&gt;, interval_left &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   interval_right &lt;chr&gt;</span></span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="prediction-of-mutation-associated-splice-targets">Prediction of mutation-associated splice targets<a class="anchor" aria-label="anchor" href="#prediction-of-mutation-associated-splice-targets"></a>
</h3>
<p>In the analysis described above we identified mutation-retrieved
splice junctions supported by RNA-seq. We have developed a stringent
detection rule to predict tumor-specific targets from
mutation-associated splice targets based on the mutation effect and the
RNA-seq support (See <span class="citation">Lang et al. (2023)</span>).
Currently, we suggest to focus on events from alternative splice sites
and exon skipping and define such splice junctions as targets that have
a |splice effect score|
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>≥</mo><annotation encoding="application/x-tex">\geq</annotation></semantics></math>
0.35 and are found by LeafCutter/SplAdder or have at least 3 junction
reads.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># exclude intron retention</span></span>
<span><span class="va">potential_targets</span> <span class="op">&lt;-</span> <span class="va">requant_noncan_mut_junctions</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">spliceai_event_type</span> <span class="op">!=</span> <span class="st">"intron retention"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">potential_targets</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">junc_id</span>, <span class="va">tx_id</span>, <span class="va">spliceai_score</span>, <span class="va">mmsplice_delta_logit_psi</span>, <span class="va">is_in_rnaseq</span>, <span class="va">junc_interval_start</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 116 × 6</span></span></span>
<span><span class="co">#&gt;    junc_id              tx_id spliceai_score mmsplice_delta_logit…¹ is_in_rnaseq</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> chr10:91528148-9152… ENST…           0.64                 <span style="color: #BB0000;">NA</span>     FALSE       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> chr10:91528148-9153… ENST…           0.22                 -<span style="color: #BB0000;">0.267</span> FALSE       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> chr10:91528148-9152… ENST…           0.64                 <span style="color: #BB0000;">NA</span>     FALSE       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> chr10:91528148-9153… ENST…           0.22                 -<span style="color: #BB0000;">0.267</span> FALSE       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> chr10:91528148-9152… ENST…           0.64                 <span style="color: #BB0000;">NA</span>     FALSE       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> chr10:91528148-9153… ENST…           0.22                 -<span style="color: #BB0000;">0.267</span> FALSE       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> chr11:18026082-1802… ENST…           0.09                 <span style="color: #BB0000;">NA</span>     FALSE       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> chr11:18026082-1802… ENST…           0.15                 -<span style="color: #BB0000;">2.50</span>  TRUE        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> chr11:18026082-1802… ENST…           0.09                 <span style="color: #BB0000;">NA</span>     FALSE       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> chr11:18026082-1802… ENST…           0.15                 -<span style="color: #BB0000;">2.50</span>  TRUE        </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 106 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ abbreviated name: ¹​mmsplice_delta_logit_psi</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: junc_interval_start &lt;dbl&gt;</span></span></span></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># apply detection rule </span></span>
<span><span class="va">targets</span> <span class="op">&lt;-</span> <span class="va">potential_targets</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">spliceai_score</span> <span class="op">&gt;=</span> <span class="fl">0.35</span> <span class="op">|</span> <span class="va">mmsplice_delta_logit_psi</span> <span class="op">&lt;</span> <span class="op">-</span><span class="fl">0.35</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">is_in_rnaseq</span> <span class="op">|</span> <span class="va">junc_interval_start</span> <span class="op">&gt;=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p>The data in <code>targets</code> can contain several altered
transcripts per junctions which may lead to several neoantigen candidate
sequences per target splice junction.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># apply detection rule </span></span>
<span><span class="va">targets</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">junc_id</span>, <span class="va">peptide_context</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">junc_id</span>, name <span class="op">=</span> <span class="st">"# neoantigen candidates"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="op">-</span><span class="va">`# neoantigen candidates`</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 7 × 2</span></span></span>
<span><span class="co">#&gt;   junc_id                   `# neoantigen candidates`</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr22:20921104-20922809:+                         3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr10:91528148-91528537:+                         2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> chr11:18026082-18029488:-                         2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> chr16:27692851-27709650:+                         2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> chr2:8910961-8916878:-                            2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> chrX:39934433-39937097:-                          2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">7</span> chrX:76944420-76953071:-                          1</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="other-splice2neo-functions">Other Splice2neo functions<a class="anchor" aria-label="anchor" href="#other-splice2neo-functions"></a>
</h2>
<div class="section level3">
<h3 id="annotation-of-transcipts">Annotation of transcipts<a class="anchor" aria-label="anchor" href="#annotation-of-transcipts"></a>
</h3>
<p>Here, we have a list of splice junctions and want to know all
potential affected transcripts.</p>
<p>Given a <code>data.frame</code> of splice junctions (with a column
<code>junc_id</code>), the function <code><a href="../reference/add_tx.html">add_tx()</a></code> annotates all
possible transcripts which overlap with the splice junctions.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">junc_df</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  junc_id <span class="op">=</span> <span class="va">toy_junc_id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">6</span>, <span class="fl">10</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">junc_df</span> <span class="op">&lt;-</span> <span class="va">junc_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/add_tx.html">add_tx</a></span><span class="op">(</span><span class="va">toy_transcripts</span><span class="op">)</span></span>
<span></span>
<span><span class="va">junc_df</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 21 × 3</span></span></span>
<span><span class="co">#&gt;    junc_id                    tx_id           tx_lst      </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;named list&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> chr2:152389996-152392205:- ENST00000409198 <span style="color: #949494;">&lt;GRanges&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> chr2:152389996-152392205:- ENST00000172853 <span style="color: #949494;">&lt;GRanges&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> chr2:152389996-152392205:- ENST00000397345 <span style="color: #949494;">&lt;GRanges&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> chr2:152389996-152392205:- ENST00000427231 <span style="color: #949494;">&lt;GRanges&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> chr2:152389996-152392205:- ENST00000618972 <span style="color: #949494;">&lt;GRanges&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> chr2:152389996-152392205:- ENST00000413693 <span style="color: #949494;">&lt;GRanges&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> chr2:152389996-152392205:- ENST00000603639 <span style="color: #949494;">&lt;GRanges&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> chr2:152389996-152392205:- ENST00000604864 <span style="color: #949494;">&lt;GRanges&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> chr2:152389996-152392205:- ENST00000420924 <span style="color: #949494;">&lt;GRanges&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> chr2:179415981-179416357:- ENST00000342992 <span style="color: #949494;">&lt;GRanges&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 11 more rows</span></span></span></code></pre></div>
<p>This can lead to large data sets containing many highly unlikely
junction transcript combinations. We can select a subset of transcripts
per junction that are more likely to be affected by a junction with
<code><a href="../reference/choose_tx.html">choose_tx()</a></code>. Please note that this function may loose
relevant or keep irrelevant junction-transcripts in particular in
regions with multiple isoforms with distinct splicing pattern.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected_junc_df</span> <span class="op">&lt;-</span> <span class="va">junc_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/choose_tx.html">choose_tx</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">selected_junc_df</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 21 × 6</span></span></span>
<span><span class="co">#&gt;    junc_id            tx_id tx_lst    putative_event_type distance_to_next_can…¹</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;named l&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> chr2:152389996-15… ENST… <span style="color: #949494;">&lt;GRanges&gt;</span> ASS                                      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> chr2:152389996-15… ENST… <span style="color: #949494;">&lt;GRanges&gt;</span> ASS                                      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> chr2:152389996-15… ENST… <span style="color: #949494;">&lt;GRanges&gt;</span> ASS                                      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> chr2:152389996-15… ENST… <span style="color: #949494;">&lt;GRanges&gt;</span> ASS                                      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> chr2:152389996-15… ENST… <span style="color: #949494;">&lt;GRanges&gt;</span> ASS                                      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> chr2:152389996-15… ENST… <span style="color: #949494;">&lt;GRanges&gt;</span> ASS                                      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> chr2:152389996-15… ENST… <span style="color: #949494;">&lt;GRanges&gt;</span> ASS                                      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> chr2:152389996-15… ENST… <span style="color: #949494;">&lt;GRanges&gt;</span> ASS                                      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> chr2:152389996-15… ENST… <span style="color: #949494;">&lt;GRanges&gt;</span> ASS                                      0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> chr2:179415981-17… ENST… <span style="color: #949494;">&lt;GRanges&gt;</span> ASS                                      0</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 11 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ abbreviated name: ¹​distance_to_next_canonical_donor</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: distance_to_next_canonical_acceptor &lt;dbl&gt;</span></span></span></code></pre></div>
<blockquote>
<p>Warning! The current implementation of <code>choose_tx</code> fails
on dataframes with many input splice junctions, likely due to a bug in a
dependency.</p>
</blockquote>
<pre><code>Error in validObject(.Object) : 
  invalid class “RangeNSBS” object: invalid object for slot "upper_bound" in class "RangeNSBS": got class "numeric", should be or extend class "integer"</code></pre>
<p>As a temporary workaround, you can group/split the dataframe by
junction and apply the method to smaller sub-dataframes. The following
code snippet uses <code>furrr::future_map</code> to parallelize the
method across multiple cores.</p>
<pre><code><span><span class="va">selected_junc_df</span> <span class="op">&lt;-</span> <span class="va">junc_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">.</span><span class="op">$</span><span class="va">junc_id</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">furrr</span><span class="fu">::</span><span class="fu">future_map</span><span class="op">(</span><span class="op">~</span><span class="fu">splice2neo</span><span class="fu">::</span><span class="fu"><a href="../reference/choose_tx.html">choose_tx</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-cheng_mmsplice_2019" class="csl-entry">
Cheng, Jun, Thi Yen Duong Nguyen, Kamil J. Cygan, Muhammed Hasan Çelik,
William G. Fairbrother, žiga Avsec, and Julien Gagneur. 2019.
<span>“<span>MMSplice</span>: Modular Modeling Improves the Predictions
of Genetic Variant Effects on Splicing.”</span> <em>Genome Biology</em>
20 (1): 48. <a href="https://doi.org/10.1186/s13059-019-1653-z" class="external-link">https://doi.org/10.1186/s13059-019-1653-z</a>.
</div>
<div id="ref-jaganathan_predicting_2019" class="csl-entry">
Jaganathan, Kishore, Sofia Kyriazopoulou Panagiotopoulou, Jeremy F.
McRae, Siavash Fazel Darbandi, David Knowles, Yang I. Li, Jack A.
Kosmicki, et al. 2019. <span>“Predicting <span>Splicing</span> from
<span>Primary</span> <span>Sequence</span> with <span>Deep</span>
<span>Learning</span>.”</span> <em>Cell</em> 176 (3): 535–548.e24. <a href="https://doi.org/10.1016/j.cell.2018.12.015" class="external-link">https://doi.org/10.1016/j.cell.2018.12.015</a>.
</div>
<div id="ref-kahles_spladder_2016" class="csl-entry">
Kahles, André, Cheng Soon Ong, Yi Zhong, and Gunnar Rätsch. 2016.
<span>“<span>SplAdder</span>: Identification, Quantification and Testing
of Alternative Splicing Events from <span>RNA</span>-<span>Seq</span>
Data.”</span> <em>Bioinformatics</em> 32 (12): 1840–47. <a href="https://doi.org/10.1093/bioinformatics/btw076" class="external-link">https://doi.org/10.1093/bioinformatics/btw076</a>.
</div>
<div id="ref-lang_identification_2022" class="csl-entry">
Lang, Franziska, Barbara Schrörs, Martin Löwer, Özlem Türeci, and Ugur
Sahin. 2022. <span>“Identification of Neoantigens for Individualized
Therapeutic Cancer Vaccines.”</span> <em>Nature Reviews Drug
Discovery</em>, February, 1–22. <a href="https://doi.org/10.1038/s41573-021-00387-y" class="external-link">https://doi.org/10.1038/s41573-021-00387-y</a>.
</div>
<div id="ref-lang_prediction_2023" class="csl-entry">
Lang, Franziska, Patrick Sorn, Martin Suchan, Alina Henrich, Christian
Albrecht, Nina Koehl, Aline Beicht, et al. 2023. <span>“Prediction of
Tumor-Specific Splicing from Somatic Mutations as a Source of Neoantigen
Candidates.”</span> bioRxiv. <a href="https://doi.org/10.1101/2023.06.27.546494" class="external-link">https://doi.org/10.1101/2023.06.27.546494</a>.
</div>
<div id="ref-li_annotation-free_2018" class="csl-entry">
Li, Yang I., David A. Knowles, Jack Humphrey, Alvaro N. Barbeira, Scott
P. Dickinson, Hae Kyung Im, and Jonathan K. Pritchard. 2018.
<span>“Annotation-Free Quantification of <span>RNA</span> Splicing Using
<span>LeafCutter</span>.”</span> <em>Nature Genetics</em> 50 (1):
151–58. <a href="https://doi.org/10.1038/s41588-017-0004-9" class="external-link">https://doi.org/10.1038/s41588-017-0004-9</a>.
</div>
<div id="ref-zeng_predicting_2022" class="csl-entry">
Zeng, Tony, and Yang I. Li. 2022. <span>“Predicting <span>RNA</span>
Splicing from <span>DNA</span> Sequence Using
<span>Pangolin</span>.”</span> <em>Genome Biology</em> 23 (1): 103. <a href="https://doi.org/10.1186/s13059-022-02664-4" class="external-link">https://doi.org/10.1186/s13059-022-02664-4</a>.
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] dplyr_1.1.4       splice2neo_0.6.14</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] DBI_1.2.3                         bitops_1.0-9                     </span></span>
<span><span class="co">#&gt;   [3] httr2_1.2.1                       permute_0.9-8                    </span></span>
<span><span class="co">#&gt;   [5] biomaRt_2.64.0                    rlang_1.1.6                      </span></span>
<span><span class="co">#&gt;   [7] magrittr_2.0.4                    vcfR_1.15.0                      </span></span>
<span><span class="co">#&gt;   [9] matrixStats_1.5.0                 compiler_4.5.1                   </span></span>
<span><span class="co">#&gt;  [11] RSQLite_2.4.3                     mgcv_1.9-3                       </span></span>
<span><span class="co">#&gt;  [13] GenomicFeatures_1.60.0            png_0.1-8                        </span></span>
<span><span class="co">#&gt;  [15] systemfonts_1.3.1                 vctrs_0.6.5                      </span></span>
<span><span class="co">#&gt;  [17] txdbmaker_1.4.2                   stringr_1.5.2                    </span></span>
<span><span class="co">#&gt;  [19] memuse_4.2-3                      pkgconfig_2.0.3                  </span></span>
<span><span class="co">#&gt;  [21] crayon_1.5.3                      fastmap_1.2.0                    </span></span>
<span><span class="co">#&gt;  [23] dbplyr_2.5.1                      XVector_0.48.0                   </span></span>
<span><span class="co">#&gt;  [25] utf8_1.2.6                        Rsamtools_2.24.1                 </span></span>
<span><span class="co">#&gt;  [27] rmarkdown_2.30                    tzdb_0.5.0                       </span></span>
<span><span class="co">#&gt;  [29] UCSC.utils_1.4.0                  ragg_1.5.0                       </span></span>
<span><span class="co">#&gt;  [31] purrr_1.1.0                       bit_4.6.0                        </span></span>
<span><span class="co">#&gt;  [33] xfun_0.53                         cachem_1.1.0                     </span></span>
<span><span class="co">#&gt;  [35] GenomeInfoDb_1.44.3               jsonlite_2.0.0                   </span></span>
<span><span class="co">#&gt;  [37] progress_1.2.3                    blob_1.2.4                       </span></span>
<span><span class="co">#&gt;  [39] DelayedArray_0.34.1               BiocParallel_1.42.2              </span></span>
<span><span class="co">#&gt;  [41] parallel_4.5.1                    prettyunits_1.2.0                </span></span>
<span><span class="co">#&gt;  [43] cluster_2.1.8.1                   R6_2.6.1                         </span></span>
<span><span class="co">#&gt;  [45] bslib_0.9.0                       stringi_1.8.7                    </span></span>
<span><span class="co">#&gt;  [47] rtracklayer_1.68.0                GenomicRanges_1.60.0             </span></span>
<span><span class="co">#&gt;  [49] jquerylib_0.1.4                   Rcpp_1.1.0                       </span></span>
<span><span class="co">#&gt;  [51] SummarizedExperiment_1.38.1       knitr_1.50                       </span></span>
<span><span class="co">#&gt;  [53] readr_2.1.5                       IRanges_2.42.0                   </span></span>
<span><span class="co">#&gt;  [55] splines_4.5.1                     Matrix_1.7-3                     </span></span>
<span><span class="co">#&gt;  [57] tidyselect_1.2.1                  abind_1.4-8                      </span></span>
<span><span class="co">#&gt;  [59] yaml_2.3.10                       vegan_2.7-2                      </span></span>
<span><span class="co">#&gt;  [61] codetools_0.2-20                  curl_7.0.0                       </span></span>
<span><span class="co">#&gt;  [63] lattice_0.22-7                    tibble_3.3.0                     </span></span>
<span><span class="co">#&gt;  [65] Biobase_2.68.0                    withr_3.0.2                      </span></span>
<span><span class="co">#&gt;  [67] KEGGREST_1.48.1                   evaluate_1.0.5                   </span></span>
<span><span class="co">#&gt;  [69] pinfsc50_1.3.0                    desc_1.4.3                       </span></span>
<span><span class="co">#&gt;  [71] BiocFileCache_2.16.2              xml2_1.4.1                       </span></span>
<span><span class="co">#&gt;  [73] Biostrings_2.76.0                 pillar_1.11.1                    </span></span>
<span><span class="co">#&gt;  [75] filelock_1.0.3                    MatrixGenerics_1.20.0            </span></span>
<span><span class="co">#&gt;  [77] stats4_4.5.1                      generics_0.1.4                   </span></span>
<span><span class="co">#&gt;  [79] vroom_1.6.6                       RCurl_1.98-1.17                  </span></span>
<span><span class="co">#&gt;  [81] S4Vectors_0.46.0                  hms_1.1.4                        </span></span>
<span><span class="co">#&gt;  [83] glue_1.8.0                        tools_4.5.1                      </span></span>
<span><span class="co">#&gt;  [85] BiocIO_1.18.0                     BSgenome_1.76.0                  </span></span>
<span><span class="co">#&gt;  [87] GenomicAlignments_1.44.0          fs_1.6.6                         </span></span>
<span><span class="co">#&gt;  [89] XML_3.99-0.19                     grid_4.5.1                       </span></span>
<span><span class="co">#&gt;  [91] tidyr_1.3.1                       ape_5.8-1                        </span></span>
<span><span class="co">#&gt;  [93] AnnotationDbi_1.70.0              nlme_3.1-168                     </span></span>
<span><span class="co">#&gt;  [95] GenomeInfoDbData_1.2.14           restfulr_0.0.16                  </span></span>
<span><span class="co">#&gt;  [97] cli_3.6.5                         rappdirs_0.3.3                   </span></span>
<span><span class="co">#&gt;  [99] textshaping_1.0.4                 viridisLite_0.4.2                </span></span>
<span><span class="co">#&gt; [101] S4Arrays_1.8.1                    BSgenome.Hsapiens.UCSC.hg19_1.4.3</span></span>
<span><span class="co">#&gt; [103] sass_0.4.10                       digest_0.6.37                    </span></span>
<span><span class="co">#&gt; [105] BiocGenerics_0.54.1               SparseArray_1.8.1                </span></span>
<span><span class="co">#&gt; [107] rjson_0.2.23                      htmlwidgets_1.6.4                </span></span>
<span><span class="co">#&gt; [109] memoise_2.0.1                     htmltools_0.5.8.1                </span></span>
<span><span class="co">#&gt; [111] pkgdown_2.1.3                     lifecycle_1.0.4                  </span></span>
<span><span class="co">#&gt; [113] httr_1.4.7                        bit64_4.6.0-1                    </span></span>
<span><span class="co">#&gt; [115] MASS_7.3-65</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jonas Ibn-Salem, Franziska Lang, Michelle Cervante, Johannes Hausmann, Patrick Sorn.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
