
# define global variables just to avoid notes in R CMD check
# See https://community.rstudio.com/t/how-to-solve-no-visible-binding-for-global-variable-note/28887

utils::globalVariables(
  c(
    ".",
    "as",
    "ALT",
    "AS_event_ID",
    "CHROM",
    "DP",
    "DP_DL",
    "DS",
    "DS_AG",
    "Gene",
    "ID",
    "POS",
    "REF",
    "SpliceAI",
    "effect",
    "chr",
    "chromosome",
    "class_mmsplice",
    "class_spliceai",
    "contig",
    "cts_id",
    "cts_junc_pos",
    "cts_seq",
    "delta_logit_psi",
    "delta_logit_psi_mmsplice",
    "downstream_end",
    "downstream_start",
    "effect",
    "efficiency",
    "efficiency_mmsplice",
    "end",
    "event_id",
    "exon1_end",
    "exon1_start",
    "exon2_end",
    "exon2_start",
    "exon_aft_start",
    "exon_alt1_end",
    "exon_alt1_start",
    "exon_alt2_end",
    "exon_alt2_start",
    "exon_const_end",
    "exon_const_start",
    "exon_end",
    "exon_id",
    "exon_idx",
    "exon_pre_end",
    "exon_start",
    "fields",
    "gene_name",
    "identified_by_leafcutter",
    "identified_by_mmsplice",
    "identified_by_spladder",
    "identified_by_spliceai",
    "incl_junc",
    "intron_cluster",
    "intron_end",
    "intron_start",
    "junc_id",
    "junc_id.x",
    "junc_id.y",
    "junc_id1",
    "junc_id2",
    "junc_id3",
    "junc_id_lst",
    "junc_tx_id",
    "junction_end",
    "junction_start",
    "junction_type",
    "left",
    "mut_effect_id",
    "mut_id",
    "mut_id.x",
    "mut_id.y",
    "pathogenicity",
    "pathogenicity_mmsplice",
    "pos_rel",
    "prob",
    "prob_spliceai",
    "queryHits",
    "right",
    "rule_left",
    "rule_right",
    "skip_junc",
    "start",
    "strand",
    "subjectHits",
    "transcript_id",
    "tx_chr",
    "tx_gr",
    "tx_id",
    "tx_id.x",
    "tx_id.y",
    "tx_name",
    "tx_nr",
    "tx_strand",
    "upstream_end",
    "upstream_start",
    "var_nr"
  )
)
