
# define global variables just to avoid notes in R CMD check
# See https://community.rstudio.com/t/how-to-solve-no-visible-binding-for-global-variable-note/28887

utils::globalVariables(
  c(
    ".",
    "as",
    "ALT",
    "AS_event_ID",
    "CHROM",
    "DP",
    "DP_DL",
    "DS",
    "DS_AG",
    "Gene",
    "ID",
    "POS",
    "REF",
    "SpliceAI",
    "effect",
    "chr",
    "chromosome",
    "class_mmsplice",
    "class_spliceai",
    "contig",
    "cts_id",
    "cts_junc_pos",
    "cts_seq",
    "delta_logit_psi",
    "delta_logit_psi_mmsplice",
    "downstream_end",
    "downstream_start",
    "efficiency",
    "efficiency_mmsplice",
    "end",
    "event_id",
    "exon1_end",
    "exon1_start",
    "exon2_end",
    "exon2_start",
    "exon_aft_start",
    "exon_alt1_end",
    "exon_alt1_start",
    "exon_alt2_end",
    "exon_alt2_start",
    "exon_const_end",
    "exon_const_start",
    "exon_end",
    "exon_id",
    "exon_idx",
    "exon_pre_end",
    "exon_start",
    "fields",
    "gene_name",
    "identified_by_leafcutter",
    "identified_by_mmsplice",
    "identified_by_spladder",
    "identified_by_spliceai",
    "incl_junc",
    "intron_cluster",
    "intron_end",
    "intron_start",
    "junc_id",
    "junc_id.x",
    "junc_id.y",
    "junc_id1",
    "junc_id2",
    "junc_id3",
    "junc_id_lst",
    "junc_tx_id",
    "junction_end",
    "junction_start",
    "junction_type",
    "left",
    "left_distance",
    "mut_effect_id",
    "mut_id",
    "mut_id.x",
    "mut_id.y",
    "pathogenicity",
    "pathogenicity_mmsplice",
    "pos_rel",
    "prob",
    "prob_spliceai",
    "queryHits",
    "right",
    "rule_left",
    "rule_right",
    "skip_junc",
    "start",
    "strand",
    "subjectHits",
    "transcript_id",
    "tx_chr",
    "tx_gr",
    "tx_id",
    "tx_id.x",
    "tx_id.y",
    "tx_name",
    "tx_nr",
    "tx_strand",
    "upstream_end",
    "upstream_start",
    "var_nr",

    "additional_novel_AA",
    "anchor",
    "annot",
    "annot_gene",
    "any_pos_on_intron",
    "at_end",
    "at_start",
    "both_pos_not_on_exon_boundaries",
    "both_pos_on_exon_boundaries",
    "both_pos_on_same_exon",
    "cds_description",
    "Chr",
    "chr_junc",
    "chrm",
    "chrom",
    "CI-SpliceAI",
    "cluster",
    "coord_1",
    "coord_2",
    "coverage_mean",
    "coverage_mean_1",
    "coverage_mean_2",
    "coverage_mean_3",
    "coverage_mean_left",
    "coverage_mean_right",
    "coverage_median",
    "coverage_median_1",
    "coverage_median_2",
    "coverage_median_3",
    "coverage_median_left",
    "coverage_median_right",
    "coverage_perc",
    "coverage_perc_1",
    "coverage_perc_2",
    "coverage_perc_3",
    "coverage_perc_left",
    "coverage_perc_right",
    "downstream_end_idx",
    "downstream_start_idx",
    "e1_end",
    "e2_end",
    "e2_start",
    "e3_end",
    "e3_start",
    "e4_start",
    "effect_index",
    "End",
    "end_junc",
    "ends_gr",
    "event_type",
    "exon_ends_lst",
    "exon_starts_lst",
    "exon_end_AA",
    "exon_end_AA_WT",
    "exons",
    "fields_split",
    "frame_shift",
    "gene",
    "gene_id",
    "gene_id_mapped",
    "gene_ids",
    "identified_by_regtools",
    "interv",
    "interval",
    "interval_1",
    "interval_2",
    "interval_3",
    "interval_exon_overlap",
    "interval_left",
    "interval_range",
    "interval_right",
    "intron",
    "intron_retention",
    "IntronDepth",
    "IRratio",
    "junc_end",
    "junc_id4",
    "junc_in_orf",
    "junc_interval_end",
    "junc_interval_start",
    "junc_start",
    "junc1",
    "junc2",
    "junc3",
    "junc4",
    "junction_start_end",
    "multi_mapping_reads",
    "name",
    "Name",
    "new_end",
    "new_start",
    "normalized_cds_junc_pos",
    "normalized_protein_junc_pos_pre",
    "normalized_protein_junc_pos",
    "number_supporting_reads",
    "other_position",
    "overlap_interval_end_reads",
    "overlap_interval_end_reads_1",
    "overlap_interval_end_reads_2",
    "overlap_interval_end_reads_3",
    "overlapping_feature",
    "Pangolin",
    "pangolin_score",
    "pep_end",
    "pep_start",
    "peptide_context",
    "peptide_context_junc_pos",
    "peptide_context_seq_raw",
    "pos",
    "pos_left_lst",
    "pos_left_on_exon",
    "pos_left_on_exon_boundary",
    "pos_left_on_exon_end",
    "pos_left_on_exon_idx",
    "pos_right_lst",
    "pos_right_on_exon",
    "pos_right_on_exon_boundary",
    "pos_right_on_exon_idx",
    "pos_right_on_exon_start",
    "protein",
    "protein_len",
    "protein_length_difference",
    "protein_until_junction",
    "protein_until_stop_codon",
    "protein_wt",
    "putative_event_type",
    "right_distance",
    "score",
    "span_interval_end",
    "span_interval_end_pairs",
    "span_interval_end_pairs_1",
    "span_interval_end_pairs_2",
    "span_interval_end_pairs_3",
    "span_interval_start",
    "splice_sites",
    "Start",
    "start_end",
    "start_junc",
    "start-end",
    "starts_gr",
    "stop_position",
    "Strand",
    "strand_junc",
    "supporting_reads_junc1",
    "supporting_reads_junc2",
    "supporting_reads_junc3",
    "supporting_reads_junc4",
    "tail",
    "tool",
    "truncated_cds",
    "tx_lst",
    "type",
    "typeCount",
    "uniquely_mapping_reads",
    "upstream_end_idx",
    "upstream_start_idx",
    "Warnings",
    "within_interval",
    "within_interval_1",
    "within_interval_2",
    "within_interval_3",
    "within_interval_left",
    "within_interval_right",
    "WT_protein_length_difference",
    "junc_id_lifted_lst",
    "pos_left_retroelement",
    "pos_left_retroelement_class",
    "pos_right_retroelement",
    "pos_right_retroelement_class",
    "left_side_retroelement",
    "right_side_retroelement"
  )
)

# This function will never be called it is just to supress NOTEs in R CMD check
# See https://r-pkgs.org/dependencies-in-practice.html#how-to-not-use-a-package-in-imports
ignore_unused_imports <- function() {
  liftOver::cur
}
